<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
        <title>M4_UserManual_5May2023_ChangesSaved</title>
        <meta name="author" content="Fleming, Sean - FPAC-NRCS, Portland, OR">
        <style type="text/css"> * {margin:0; padding:0; text-indent:0; }
 h1 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 28pt; }
 h2 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 18pt; }
 .s1 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 11.5pt; vertical-align: 5pt; }
 .s2 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 11pt; }
 h3 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 16pt; }
 .s3 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 10.5pt; vertical-align: 5pt; }
 .p, p { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; margin:0pt; }
 .s4 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 7pt; vertical-align: 3pt; }
 h4 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 14pt; }
 .s5 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 .s6 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 8pt; vertical-align: 4pt; }
 .s7 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 9pt; vertical-align: 4pt; }
 .s8 { color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 8pt; }
 .s9 { color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 8pt; }
 .s10 { color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 10pt; }
 .s11 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 7pt; vertical-align: 3pt; }
 .s13 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 2pt; }
 .s14 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 12pt; }
 .s15 { color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 8pt; vertical-align: 4pt; }
 .s16 { color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 .s17 { color: black; font-family:"Times New Roman", serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 .s18 { color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 9pt; }
 .s19 { color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 9pt; }
 .s20 { color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 li {display: block; }
 #l1 {padding-left: 0pt;counter-reset: c1 1; }
 #l1> li>*:first-child:before {counter-increment: c1; content: counter(c1, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 16pt; }
 #l1> li:first-child>*:first-child:before {counter-increment: c1 0;  }
 #l2 {padding-left: 0pt;counter-reset: c2 1; }
 #l2> li>*:first-child:before {counter-increment: c2; content: counter(c1, decimal)"."counter(c2, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 14pt; }
 #l2> li:first-child>*:first-child:before {counter-increment: c2 0;  }
 #l3 {padding-left: 0pt;counter-reset: d1 1; }
 #l3> li>*:first-child:before {counter-increment: d1; content: counter(d1, decimal)". "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l3> li:first-child>*:first-child:before {counter-increment: d1 0;  }
 #l4 {padding-left: 0pt;counter-reset: c3 1; }
 #l4> li>*:first-child:before {counter-increment: c3; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 #l4> li:first-child>*:first-child:before {counter-increment: c3 0;  }
 #l5 {padding-left: 0pt; }
 #l5> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l6 {padding-left: 0pt; }
 #l6> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l7 {padding-left: 0pt; }
 #l7> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l8 {padding-left: 0pt;counter-reset: c3 1; }
 #l8> li>*:first-child:before {counter-increment: c3; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 #l8> li:first-child>*:first-child:before {counter-increment: c3 0;  }
 #l9 {padding-left: 0pt;counter-reset: c2 1; }
 #l9> li>*:first-child:before {counter-increment: c2; content: counter(c1, decimal)"."counter(c2, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 14pt; }
 #l9> li:first-child>*:first-child:before {counter-increment: c2 0;  }
 #l10 {padding-left: 0pt;counter-reset: h1 1; }
 #l10> li>*:first-child:before {counter-increment: h1; content: counter(h1, decimal)". "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l10> li:first-child>*:first-child:before {counter-increment: h1 0;  }
 #l11 {padding-left: 0pt;counter-reset: c3 1; }
 #l11> li>*:first-child:before {counter-increment: c3; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 #l11> li:first-child>*:first-child:before {counter-increment: c3 0;  }
 #l12 {padding-left: 0pt; }
 #l12> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l13 {padding-left: 0pt; }
 #l13> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l14 {padding-left: 0pt; }
 #l14> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l15 {padding-left: 0pt; }
 #l15> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l16 {padding-left: 0pt;counter-reset: c4 1; }
 #l16> li>*:first-child:before {counter-increment: c4; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)"."counter(c4, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 12pt; }
 #l16> li:first-child>*:first-child:before {counter-increment: c4 0;  }
 #l17 {padding-left: 0pt; }
 #l17> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l18 {padding-left: 0pt; }
 #l18> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l19 {padding-left: 0pt; }
 #l19> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l20 {padding-left: 0pt; }
 #l20> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l21 {padding-left: 0pt; }
 #l21> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l22 {padding-left: 0pt; }
 #l22> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l23 {padding-left: 0pt; }
 #l23> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l24 {padding-left: 0pt; }
 #l24> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l25 {padding-left: 0pt; }
 #l25> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l26 {padding-left: 0pt; }
 #l26> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l27 {padding-left: 0pt; }
 #l27> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l28 {padding-left: 0pt; }
 #l28> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l29 {padding-left: 0pt; }
 #l29> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l30 {padding-left: 0pt; }
 #l30> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l31 {padding-left: 0pt; }
 #l31> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l32 {padding-left: 0pt; }
 #l32> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l33 {padding-left: 0pt; }
 #l33> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l34 {padding-left: 0pt; }
 #l34> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l35 {padding-left: 0pt; }
 #l35> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l36 {padding-left: 0pt; }
 #l36> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l37 {padding-left: 0pt; }
 #l37> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l38 {padding-left: 0pt; }
 #l38> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l39 {padding-left: 0pt; }
 #l39> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l40 {padding-left: 0pt; }
 #l40> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l41 {padding-left: 0pt; }
 #l41> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l42 {padding-left: 0pt; }
 #l42> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l43 {padding-left: 0pt; }
 #l43> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l44 {padding-left: 0pt;counter-reset: c3 1; }
 #l44> li>*:first-child:before {counter-increment: c3; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 #l44> li:first-child>*:first-child:before {counter-increment: c3 0;  }
 #l45 {padding-left: 0pt;counter-reset: c4 1; }
 #l45> li>*:first-child:before {counter-increment: c4; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)"."counter(c4, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 12pt; }
 #l45> li:first-child>*:first-child:before {counter-increment: c4 0;  }
 #l46 {padding-left: 0pt; }
 #l46> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l47 {padding-left: 0pt; }
 #l47> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l48 {padding-left: 0pt; }
 #l48> li>*:first-child:before {content: "o "; color: black; font-family:"Courier New", monospace; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l49 {padding-left: 0pt; }
 #l49> li>*:first-child:before {content: " "; color: black; font-family:Wingdings; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l50 {padding-left: 0pt;counter-reset: c4 1; }
 #l50> li>*:first-child:before {counter-increment: c4; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)"."counter(c4, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 12pt; }
 #l50> li:first-child>*:first-child:before {counter-increment: c4 0;  }
 #l51 {padding-left: 0pt; }
 #l51> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 #l52 {padding-left: 0pt;counter-reset: c2 1; }
 #l52> li>*:first-child:before {counter-increment: c2; content: counter(c1, decimal)"."counter(c2, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: normal; font-weight: bold; text-decoration: none; font-size: 14pt; }
 #l52> li:first-child>*:first-child:before {counter-increment: c2 0;  }
 #l53 {padding-left: 0pt;counter-reset: c3 1; }
 #l53> li>*:first-child:before {counter-increment: c3; content: counter(c1, decimal)"."counter(c2, decimal)"."counter(c3, decimal)" "; color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: bold; text-decoration: none; font-size: 12pt; }
 #l53> li:first-child>*:first-child:before {counter-increment: c3 0;  }
 li {display: block; }
 #l54 {padding-left: 0pt; }
 #l54> li>*:first-child:before {content: " "; color: black; font-family:Symbol, serif; font-style: normal; font-weight: normal; text-decoration: none; font-size: 11pt; }
 table, tbody {vertical-align: top; overflow: visible; }
        </style>
    </head>
    <body>
        <h1 style="padding-left: 8pt;text-indent: 0pt;line-height: 32pt;text-align: left;">User Manual</h1>
        <h2 style="padding-top: 10pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
            US Department of Agriculture Snow Survey and Water Supply Forecasting Program Multi Model Machine-learning Metasystem (M
            <span class="s1">4</span>
            )
        </h2>
        <p style="text-indent: 0pt;text-align: left;">
            <br>
        </p>
        <p class="s2" style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">First complete draft 10 March 2023, S.W. Fleming, NRCS National Water and Climate Center<br>Last modified 5 May 2023, S.W. Fleming, NRCS National Water and Climate Center</p>
        <p class="s2" style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
            <a href="mailto:beau.uriona@usda.gov" style=" color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 11pt;" target="_blank">For information contact Beau Uriona (</a>
            <span style=" color: #0562C1; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: underline; font-size: 11pt;">beau.uriona@usda.gov</span>
            <a href="mailto:sean.fleming@usda.gov" style=" color: black; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: none; font-size: 11pt;" target="_blank">) or Sean Fleming (</a>
            <span style=" color: #0562C1; font-family:Calibri, sans-serif; font-style: italic; font-weight: normal; text-decoration: underline; font-size: 11pt;">sean.fleming@usda.gov</span>
            ) at the USDA Natural Resources Conservation Service
        </p>
        <p style="text-indent: 0pt;text-align: left;">
            <br>
        </p>
        <ol id="l1">
            <li data-list-text="1">
                <h3 style="padding-left: 44pt;text-indent: -36pt;text-align: left;">
                    What is M
                    <span class="s3">4</span>
                    ?
                </h3>
                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    The multi-model machine-learning metasystem (M
                    <span class="s4">4</span>
                    ) is the prototype next-generation mathematical and software model for operational water supply forecasting (seasonal river flow volume prediction) built for and employed by the Snow Survey and Water Supply Forecasting (SSWSF) Program of the US Department of Agriculture (USDA) Natural Resources Conservation Service (NRCS), which spans close to 600 forecast locations across the American West. It will replace the linear principal component regression (PCR) method that the NRCS SSWSF Program, and in particular the National Water and Climate Center (NWCC), has used as its primary (but not sole) water supply forecasting (WSF) model in this West-wide NRCS operational role since the early 1990s. M
                    <span class="s4">4</span>
                    leverages a broadly PCR-like architecture and philosophy but augments it with multi-model ensemble modeling for improved forecast stability and accuracy, explainable and automated machine learning/artificial intelligence (ML/AI) for improved forecast accuracy, evolutionary computing to support global feature optimization, more advanced statistical methods for prediction uncertainty estimation under non-Gaussian and heteroscedastic error distributions, and other advances that were directly motivated by the day-to-day operational requirements of a governmental service-delivery organization performing operational water supply forecasting at scale. It is written in the free, open-source R scientific/statistical computing language.
                </p>
                <p style="text-indent: 0pt;text-align: left;">
                    <br>
                </p>
            </li>
            <li data-list-text="2">
                <h3 style="padding-left: 44pt;text-indent: -36pt;text-align: left;">Disclaimer</h3>
                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    The NWCC posts the M
                    <span class="s4">4</span>
                    code publicly here for four reasons: (a) to satisfy internal NRCS archiving, access, and version-control needs around the code and its documentation, including instructions for its use; (b) to demonstrate full transparency in the techniques used by the NRCS SSWSF Program, and in so doing, and in conjunction with extensive vetting of the M
                    <span class="s4">4</span>
                    methodology and performance in the peer- reviewed science and engineering literature, to support the technical credibility of the program and public confidence in its products; (c) as a practical mechanism for sharing M
                    <span class="s4">4</span>
                    with partner agencies and organizations; and (d) to facilitate and encourage further refinement and development of M
                    <span class="s4">4</span>
                    by making it available as open-source code to the STEM research community.
                </p>
                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    Any purpose to which this code is put by external users downloading M
                    <span class="s4">4</span>
                    from this site is entirely at the discretion of those users, who accept full responsibility for such uses and their consequences, which are in general beyond our knowledge or control. M
                    <span class="s4">4</span>
                    is an operational prototype, and while it has undergone successful testing, much of which has been documented in the peer-reviewed science and engineering literature, the possibility of bugs or errors cannot be ruled out (see also Section 6 below).
                </p>
                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    Note that operational forecasts officially issued by the NRCS under a specific set of user protocols and institutional controls, irrespective of whether those forecasts were or were not developed in part using guidance from M
                    <span class="s4">4</span>
                    , are distinct from the predictions and performance of the M
                    <span class="s4">4</span>
                    code itself as posted here to GitHub and for which no claims are made as to accuracy or value. Additionally, all environmental analysis and prediction software, including M
                    <span class="s4">4</span>
                    , has inherent limitations and idiosyncrasies. The characteristics of the prototype M
                    <span class="s4">4</span>
                    probabilistic nonlinear regression metasystem are such that it may be useful for purposes other than the WSF applications for which it was designed, but while we applaud exploration of such additional applications, we emphasize that absolutely no testing of any such applications has been undertaken by the NRCS and no claims as to its performance in or suitability for such applications are made. No warranties, guarantees, or any other assurances whatsoever around M
                    <span class="s4">4</span>
                    and its results in any application, WSF-related or not, are provided. Users who believe they have discovered a problem are encouraged to report it to the NRCS.
                </p>
                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    Users of M
                    <span class="s4">4</span>
                    should read and understand this user manual in its entirety, including the externally peer- reviewed journal articles in its Appendices which describe the M
                    <span class="s4">4</span>
                    prototype. Note also that while it is hoped that the current distribution of the M
                    <span class="s4">4</span>
                    prototype will be taken as a starting point for additional M
                    <span class="s4">4</span>
                    development, and there is no shortage of potential future directions that such further development might take, developers are responsible for their own results and should familiarize themselves thoroughly with the peer-reviewed literature on M
                    <span class="s4">4</span>
                    and with the code itself in its entirety: there are a lot of moving parts in M
                    <span class="s4">4</span>
                    , many of which are interconnected, especially as they relate to functionalities specifically directed to operational WSF environments and standard WSF use-cases and AutoML functionalities (see below). Commenting in the code is fairly ubiquitous, which may hopefully help scientific software developers decipher the intent and rationales of certain code snippets and the overall workflows in M
                    <span class="s4">4</span>
                    .
                </p>
                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    While just about anyone is welcome to experiment with M
                    <span class="s4">4</span>
                    , please be aware that some required but not necessarily sufficient prerequisites for using M
                    <span class="s4">4</span>
                    correctly, using its results appropriately, and accurately understanding its nature and capabilities, are good familiarity with current standard procedures in statistical WSF in the western US as implemented by NRCS and other similar operational service-delivery organizations, solid experience with R or at a minimum other similar interpreted scientific computing languages, and a thorough understanding of the principles, goals, methods, and capabilities and limitations of M
                    <span class="s4">4</span>
                    as described in the peer-reviewed literature and this manual.
                </p>
                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">The code, as it is posted here, should be viewed as experimental – use at your own risk! See also the GitHub license agreement.</p>
            </li>
            <li data-list-text="3">
                <h3 style="padding-top: 2pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                    A synopsis of M
                    <span class="s3">4</span>
                    philosophy, context, and basic workflow
                </h3>
                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                    Full explanations of M
                    <span class="s4">4</span>
                    and issues motivating its development and design choices are provided in the Appendices, which users should read and digest. The following is a short summary of the general context, philosophy, and workflows associated the new NRCS forecasting metasystem.
                </p>
                <p style="text-indent: 0pt;text-align: left;">
                    <br>
                </p>
                <ol id="l2">
                    <li data-list-text="3.1">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">Design criteria</h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            M
                            <span class="s4">4</span>
                            did not transition from research to operations (R2O) per se. Rather, a more pragmatic applied- science and engineering paradigm was used: operational requirements were first defined, and M
                            <span class="s4">4</span>
                            was then designed and built to meet those requirements. These specific design criteria for operational WSF at NRCS, many of which are applicable more generally to operational hydrologic forecasting at other service-delivery organizations, included:
                        </p>
                        <ol id="l3">
                            <li data-list-text="1.">
                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;text-align: left;">Improved forecast accuracy</p>
                            </li>
                            <li data-list-text="2.">
                                <p style="padding-top: 1pt;padding-left: 43pt;text-indent: -18pt;text-align: left;">Improved potential for automation (largely an “over-the-loop” system)</p>
                            </li>
                            <li data-list-text="3.">
                                <p style="padding-top: 1pt;padding-left: 44pt;text-indent: -18pt;text-align: left;">Relatively low cost &amp; good ease of development, implementation, and operation</p>
                            </li>
                            <li data-list-text="4.">
                                <p style="padding-top: 1pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Address known technical issues with the existing system by seamlessly accommodating nonlinear functional forms, and heteroscedastic and non-Gaussian residuals requiring time- varying &amp; asymmetric prediction intervals, without the use of predictand transforms</p>
                            </li>
                            <li data-list-text="5.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Flexible, modular, expandable: relatively easy ability to integrate emerging new techniques as appropriate</p>
                            </li>
                            <li data-list-text="6.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Physics-aware AI: hydrologic theory-guided system, generating geophysically explainable forecasts</p>
                            </li>
                            <li data-list-text="7.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Multi-model ensemble paradigm: address equifinality and model selection uncertainty, with robustness over diverse geophysical environments across the US West and Alaska</p>
                            </li>
                            <li data-list-text="8.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Accommodate high-dimensional predictor datasets &amp; potential for multiple independent input signals: will only grow more important with increasing use of highly spatially distributed input datasets such as those from remote sensing or climate modeling products</p>
                            </li>
                            <li data-list-text="9.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Balance innovation and performance gains vs. established building blocks and proven tools: apply most advanced yet acceptable (MAYA) industrial design principle.</p>
                            </li>
                        </ol>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="3.2">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">MAYA</h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">The MAYA principle was introduced by the renown French-American industrial designer, Raymond Loewy. In the context of a next-generation WSF system, MAYA signifies a compromise between adopting the most advanced techniques coming out of the research community, which offer great promise but remain largely unproven in real-world high-stakes applications having strong organizational accountabilities, and established technologies and practices that have proven themselves and achieved buy-in from the water resource science, engineering, and management communities but may not represent the best-performing methods now available. The goal is a unique new mission-specific prediction model, using a carefully selected suite of up-to-date methods from the geophysical modeling</p>
                        <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">and data science communities, yet one that also corresponds to a ‘hot-rodded’ version of the proven and accepted PCR-based approach to WSF.</p>
                        <ol id="l4">
                            <li data-list-text="3.2.1">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                                    Synopsis of some conventional elements of M
                                    <span class="s6">4</span>
                                </p>
                                <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The conventional elements of M
                                    <span class="s4">4</span>
                                    include a PCR-like structure, adoption of the canonical use-case for statistical WSF in the western US, quantitative estimation of forecast uncertainty in the form of prediction intervals, incorporation of individually well-proven techniques, and an emphasis on the ability to easily generate a relatable storyline for any given forecast on any given day.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Linear PCR was introduced to operational WSF by the NRCS in the early 1990s as a replacement for earlier linear regression methods. Its primary advantage is that it provides a theoretical and practical framework for performing linear regression with multicollinear predictors. The basic workflow of PCR involves performing a principal component analysis (PCA) on the input data matrix, and using one or more of the resulting modes, starting with the leading mode, as predictors in an otherwise standard linear regression to predict seasonal river flow volume. Because inputs are mainly measurements of snow water equivalent (SWE) and other similar hydroclimatic parameters, the leading PCA mode effectively serves as an index of watershed-scale hydroclimatic conditions. Higher modes may be retained, reflecting more subtle hydroclimatic patterns, or capturing other processes like aquifer-stream interactions if antecedent streamflow is additionally included as a predictor. Out-of-sample prediction performance is normally found as the leave-one-out cross-validated (LOOCV; sometimes called jackknifed though strictly speaking this refers to something slightly different) standard error, root mean square error, coefficient of determination, or similar metric. This LOOCV procedure is standard practice in statistical WSF systems deployed operationally by service-delivery organizations in western North America, and larger cross-validation folds are generally not required because, while many hydroclimatic variables are serially correlated, the residuals from WSF regression models typically demonstrate little or no statistically significant autocorrelation at the annual timescale considered in the canonical western US statistical WSF use-case (see below; see also examples given later in this manual, and in the M
                                    <span class="s4">4</span>
                                    distribution files; the appendices also provide more information). That cross-validated performance measure then provides the basis for forming prediction intervals, expressed for example in NRCS practice as 90%, 70%, 30%, and 10% exceedance flows. In particular, a simple but common heuristic is used: the probability density function for model error is taken to be a stationary Gaussian (normal) distribution, with a mean equal to that deterministic prediction and a standard deviation equal to the LOOCV standard error or more-or-less equivalently the root mean square error, and the exceedance flows are estimated as the 0.10, 0.30, 0.70, and 0.90 quantiles of that distribution. The entire workflow may be wrapped in an automated system for identifying optimal predictors and PCA modes to retain.
                                </p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The foregoing processes have been implemented in many software systems over the years, including the VIPER program used operationally by NRCS. This proven overall procedure was in many, but not all, respects largely replicated in M
                                    <span class="s4">4</span>
                                    , which augments it with the use of multi-model ensemble modeling, a genetic algorithm for optimal feature extraction, and more sophisticated and flexible methods for estimating prediction uncertainty.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">The canonical use-case for statistical WSF in the western US involves a single predictand, which is the upcoming spring-summer flow volume at a given location on a given river; predictors consisting mainly of current SWE and year-to-date precipitation at various locations in and near the watershed, and possibly antecedent streamflow or an ENSO index; and a different model for each forecast point-</p>
                                <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">publication date-target period triplet. To elaborate, operational statistical WSF models at practical service-delivery organizations (SDOs) having strong accountabilities for reliably producing and publicly disseminating forecast information are organized around solving a specific forecasting problem (SFP). An SFP is a unique combination of annual publication date, target period, and forecast point.</p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">Publication date is the day of the year that the forecast is issued. Common examples are January 1, February 1, March 1, April 1, May 1, or June1, but others are possible depending on the particular river, reflecting local hydroclimatic characteristics and water management concerns. The target period defines the upcoming seasonal timeframe for which accumulated river runoff volume is being forecasted; a prevalent example is April-July but others are commonly used as well, again depending on the particular river, reflecting local hydroclimatic characteristics and water management concerns. The forecast point is a particular location on a particular river for which a runoff forecast is being made.</p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    There may be more than one forecast point on a given river, especially long rivers; but each forecast point corresponds to a separate SFP. An example of a SFP is a water supply forecast, issued on February 1 (publication date), of the total April-July flow volume (target period) of the Deschutes River below Snow Creek (forecast point). A unique model is built for each SFP. However, there may be a great deal of overlap between models for similar SFPs. For example, the January 1 and April 1 publication-date models for the Deschutes River below Snow Creek may use updated versions of the same predictors, such as SWE at some SNOTEL station on January 1 and on April 1, respectively. However, the relevance of certain predictors may also change over the course of the forecast season reflecting seasonal hydroclimatic processes, so the optimal predictor set may gradually change as well; for example, ENSO indices, in regions where they are useful, tend to provide much of their value relatively early in the forecast season when they provide some information about upcoming winter-spring rainfall and snowfall. M
                                    <span class="s4">4</span>
                                    was built for this canonical use-case, albeit with an eye to make better use of emerging predictors.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Western water management can be a very high-stakes enterprise, and the water managed using WSFs may have values ranging into the billions or trillions of dollars every year. While this militates on the one hand for using the best available prediction technologies, responsibilities around sound decision- making, and the economic and political ramifications of making bad choices, also call for a certain level of prudence and conservatism. This risk-aversion leads to a reliance on proven methods. M
                                    <span class="s4">4</span>
                                    employs a proven PCR-like architecture and adheres to the canonical WSF use-case, as described above, and its constituent elements consist of methods that have individually been widely tested and accepted in physical, statistical, and machine learning applications.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Similarly, most water resource scientists, engineers, and managers hold low trust in predictions that are not geophysically explainable. A readily identifiable and relatable ‘storyline’ for any given forecast on any given day is, in practice, a necessary ingredient of any operational WSF modeling system at a SDO. This typically consists of a general explanation of why a certain forecast was obtained, in light of well- understood terrestrial hydrologic processes and current watershed-scale climatic (snowpack, soil moisture, etc.) conditions. Producing such a physical explanation of the forecast does not necessarily require a process simulation-based hydrologic model. Use of PCR by well-trained operational hydrologists, for example, has produced clear hydroclimatic storylines for decades. The ability to continue this forward into the machine-learning world was one of the most important and challenging design goals for M
                                    <span class="s4">4</span>
                                    .
                                </p>
                            </li>
                            <li data-list-text="3.2.2">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Synopsis of some fresh elements</p>
                                <p style="padding-top: 3pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    New elements in M
                                    <span class="s4">4</span>
                                    include a multi-model ensemble forecasting framework, the first wide-scale incorporation of AI into a governmental SDO operational WSF system, evolutionary computing, and advanced statistical prediction uncertainty estimation.
                                </p>
                                <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The final product of M
                                    <span class="s4">4</span>
                                    is a multi-model ensemble mean forecast. M
                                    <span class="s4">4</span>
                                    consists of six semi-independent forecast modeling systems acting in parallel and may be viewed as a system of systems (hence ‘metasystem’). Each of the constituent modeling systems has a PCR-like architecture as noted above but replaces simple linear regression and prediction intervals based on stationary normal error distributions with various advanced statistical or machine learning algorithms. The results of all six are averaged together to create the ensemble mean forecast. Advantages of this approach are that it deals with model equifinality, that is, similar overall performance from several quite different models, and provides more stable, and in some cases, more accurate results than any of the individual models alone. It is a standard operational approach in numerical weather prediction and climate modeling, it is very well- established in applied hydrologic research using process-based models, and it is used within some machine learning approaches, such as bootstrap aggregation (bagging) within some neural networks.
                                </p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">However, multi-model ensembles have not, in general, migrated into operational river forecasting systems at SDOs (although related but simpler concepts like extended (or ensemble) streamflow prediction (ESPs), which use a single hydrologic model with multiple climate traces, made the jump decades ago). We emphasize that using the results of a multi-model hydrologic prediction system wisely can require operating in a different headspace than some operational hydrologists may be accustomed to. For example, hunting-and-pecking for which of the six constituent models works ‘best’ for some particular combination of forecast point, publication date, and target period misses the point of the exercise and can lead to wasted time and ultimately, overall, an underperforming modeling strategy.</p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">More background on this sometimes-counterintuitive approach can be found in the Appendices, and the publications cited in the Appendices.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">Research using AI for river forecasting dates back to 1995 or possibly earlier, and its superior prediction accuracy has been proven repeatedly. Mainstreaming of AI into wider and operational hydrologic applications has been much more recent and limited due to various stumbling blocks. These included a lack of prediction uncertainty intervals; a lack of explainability, i.e., the so-called black-box problem; skepticism in high-stakes operational settings where existing methods are well-established and new methods coming out of the research community are often viewed as lacking real-world relevance and credibility; and misunderstandings and lack of technical training and professional familiarity among many physical scientists in both the research and operational communities regarding AI and its real limitations and strengths. But things are changing, as almost everyone is getting used to the idea of AI, machine learning tools have become far more accessible and methods are far more diversified, and explainability, overtraining, prediction uncertainty estimation, and other questions can be addressed with appropriate AI design and implementation choices. WSF applications of AI, however, still have to be approached carefully and with the goal of satisfying specific user needs, rather than throwing machine learning at a problem and expecting the result to be useful and accepted.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Accomplishing this, as noted above, was a key goal when developing M
                                    <span class="s4">4</span>
                                    . The approach and results are documented in the Appendices, and some cursory summaries are provided in the following subsections.
                                </p>
                                <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">Before forging ahead, it’s probably also useful to provide a glossary of some basic machine learning jargon, bearing in mind however that these terms are evolving quickly and may be used in slightly different ways in different fields:</p>
                                <ul id="l5">
                                    <li data-list-text="">
                                        <p class="s2" style="padding-top: 7pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Data science
                                            <span class="p">Algorithms for building data-based information/knowledge pipelines, often to enable automated predictions and actions; spans traditional statistical modeling and AI</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Data mining
                                            <span class="p">Extracting descriptive, explanatory, or predictive patterns from data, typically without clear a priori expectations around specific causal relationships</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Big data
                                            <span class="p">Satisfy three Vs of volume, velocity, and variety: gigantic and never-ending torrents of data having a wide &amp; unpredictable content range (e.g., YouTube uploads). Strictly speaking, few environmental datasets satisfy this definition, but big-data methods can nevertheless be useful. A fourth V – veracity – is also sometimes included.</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Artificial intelligence (AI)
                                            <span class="p">Technologies that emulate human intelligence in various ways, in turn including several subdisciplines, like machine learning and robotics</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Machine learning (ML)
                                            <span class="p">An AI field using algorithms to identify patterns in data and, typically, applying to those patterns to make predictions; divided into classification or regression, and unsupervised or supervised methods; examples include a wide variety of artificial neural network (ANN) types, random forests (RFs), support vector machines (SVMs), and deep learning (DL) architectures like long short term memory (LSTM) networks and convolutional neural networks (CNNs); these algorithms are often inspired by biological systems, such as the human brain (ANNs) or human decision-making processes (the decision trees in RFs)</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Hyperparameters
                                            <span class="p">AIs have parameters loosely analogous to coefficients in a linear statistical model (e.g., a neural network’s neuron weights) that are optimized in training; but AIs additionally have higher-level hyperparameters (e.g., a neural network’s learning rate) that control this process of estimating parameter values and the overall architecture</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Features and targets
                                            <span class="p">Akin to predictors and predictands, respectively, in traditional statistical modeling; features engineering is the procedure of processing/manipulating input data to extract and select features to be passed to the AI as predictor variables, a major element of AI application; one of the goals of deep learning is to integrate features engineering into the core of the machine learning algorithm, avoiding the need to perform features engineering explicitly; conversely, user-directed features engineering has been identified as a type of physics-informed machine learning</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            AutoML
                                            <span class="p">System to automatically build the best AI for a given dataset; may include optimal hyperparameter selection and is intended to make ML easier for non-AI specialists to apply in respective practice domains; also creates the need for still-higher-level ‘hyperhyperparameters’</span>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Overtraining and regularization
                                            <span class="p">Overtraining is memorization of the training data, compromising generalization accuracy; it happens in all models involving calibration or derivation of parameters from observational data, not just ML, including physics-based and conventional linear statistical models, but it can be particularly acute for ML due the tremendous fitting ability of machine learning algorithms; however, it is easily detectable in validation; regularization refers to technical methods that mitigate it.</span>
                                        </p>
                                    </li>
                                </ul>
                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Evolutionary computing is also used in M
                                    <span class="s4">4</span>
                                    . This refers to a family of computing techniques that emulate biological evolutionary processes. M
                                    <span class="s4">4</span>
                                    deploys a genetic algorithm, which is a type of optimization procedure, for optimizing the features (specifically, the combination of input predictor variables and
                                </p>
                                <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCA modes) fed to its constituent modeling systems. The goal is identical to techniques often conventionally used for optimizing these choices, such as the tree algorithm used in the NRCS VIPER implementation of PCR, but the methods it uses are more advanced. Of particular note is that it is a stochastic optimization scheme, which helps avoids having the solution process get trapped in local minima in a complex nonlinear error space, contributing to a greater likelihood of obtaining a global (true) optimum.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Considered in aggregate, the foregoing attributes mean that M
                                    <span class="s4">4</span>
                                    can be viewed as a community of artificial lifeforms. Each of these AIs consists of an artificial intelligence, and following rules adapted from genetics, each evolves over a series of generations in such a way as to perform the best it can in its environment. Ultimately, this community of six artificial lifeforms reaches a consensus decision on the best WSF and how much certainty it collectively has in that decision.
                                </p>
                                <p style="text-indent: 0pt;text-align: left;">
                                    <br>
                                </p>
                            </li>
                        </ol>
                    </li>
                    <li data-list-text="3.3">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">Explainable machine learning</h4>
                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            Several pragmatic, somewhat WSF-specific steps were taken in M
                            <span class="s4">4</span>
                            design to dovetail machine learning with the underlying process physics of watershed hydrology:
                        </p>
                        <ul id="l6">
                            <li data-list-text="">
                                <p class="s2" style="padding-top: 7pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    Operational hydrologist-based features engineering
                                    <span class="p">User-directed selection of predictive data, and how that data are compressed and fed to supervised learning algorithms, reflects end-user knowledge around representativeness, reliability, quirks, and capabilities of potential input variables and measurement sites and geophysical interpretations of PCA modes. That the hydrologist should pick what data to use and how to process it may seem obvious, but it is somewhat of a counterpoint to some big data-focused deep learning approaches, where large datasets are fed to complex multi-layer neural networks that, in a nutshell, perform features engineering automatically. It is a key location in AI development process for domain experts to insert their physical hydrologic knowledge.</span>
                                </p>
                            </li>
                            <li data-list-text="">
                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    <i>A priori physicality constraints.</i>
                                    Some feature-target relationships are known to be significantly nonlinear but monotonic. For example, a high-snow winter, all other things being equal, won’t produce a low-flow spring and summer. Certain AI methods allow a monotonicity constraint to be enforced. While overall monotonicity of the ensemble mean relationship is not strictly enforced, M
                                    <span class="s4">4</span>
                                    strongly encourages that net behavior, as two of its six methods are intrinsically monotonic and another two have the option, which is activated in the default application of M
                                    <span class="s4">4</span>
                                    , of enforcing monotonicity constraints. Similarly, negative-valued volume predictions are non- physical yet can happen in some prediction systems, and certain AI methods allow a non- negativity constraint to be enforced. Several such methods are used here. Moreover, the use of nonlinear ML techniques tends to avoid the generation of negative-valued best-estimate predictions. M
                                    <span class="s4">4</span>
                                    also includes algorithmic logic to sequentially prune ensemble members if the 90% exceedance flow in the ensemble mean forecast distribution violates the nonnegativity constraint for any sample in the training interval, though this does not necessarily guarantee that a forecast going forward is strictly nonnegative. Note that monotonicity and nonnegativity constraints are simple forms of theory-guided data science, and that by severely restricting the solution space available to the MLs, such geophysical plausibility constraints improve both regularization and explainability.
                                </p>
                            </li>
                            <li data-list-text="">
                                <p class="s2" style="padding-top: 3pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    WSF is purposefully framed as low-dimensional problem with a parsimonious solution
                                    <span class="p">. PCA pre- processing dramatically reduces problem dimensionality, and the predictand is a simple scalar. Additionally, considerable effort and testing was investing in identifying default hyperparameters and AutoML algorithms that help create the most parsimonious ML architecture possible for the canonical statistical WSF use-case. The result is, in most such applications, an extremely compact ML problem with one or two inputs and a single output. The relationships between those inputs and outputs can be easily visualized in a x-y or contour plot, allowing the user to see what the ML is thinking. Another advantage of low-dimensional features and targets and compact architectures (e.g., minimizing the number of hidden-layer</span>
                                </p>
                            </li>
                        </ul>
                        <p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">neurons) is that it reduces the number of parameters that must be trained, making the ML more amenable to application using small sample sizes.</p>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="3.4">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">AutoML</h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            While completely hands-off operation of a WSF system is in general not desired by SDOs, streamlining and automation of many tasks is, both in the interest of operational efficiency and to make a system like M
                            <span class="s4">4</span>
                            accessible to users who expertise lies in areas other than AI. This leads to judicious use of automatic, or autonomous, machine learning (commonly dubbed AutoML). Its use dovetails with “over-the-loop” hydrometeorological forecasting concepts. In particular:
                        </p>
                        <ul id="l7">
                            <li data-list-text="">
                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">Algorithmic logic is introduced to automate most optimization/decision points (including ML hyperparameters). However, manual overrides remain available.</p>
                            </li>
                            <li data-list-text="">
                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">Other hyperparameters are set to robust default values on the basis of extensive experimentation using various NRCS WSF test cases and problem setups. For example, hundreds of metasystem runs were completed to locate values for the population size and the number of generations in the genetic algorithm that work as generally serviceable defaults for the canonical statistical WSF use-case.</p>
                            </li>
                        </ul>
                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">The ability to override these defaults and AutoML algorithms is also considered important at SDOs, providing hydrologists with an opportunity to make adjustments in accordance with their professional judgement.</p>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="3.5">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">
                            M
                            <span class="s7">4</span>
                            development approach, process, and status
                        </h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            Why choose R? There were four reasons. First, it’s free for everyone. Second, the design criteria for M
                            <span class="s4">4</span>
                            included some very specific requirements around a priori physicality constraints and non-standard statistical properties of the forecast distributions, and to our knowledge, these are most readily, and perhaps exclusively, available in certain existing R packages previously written by others and freely available on the CRAN server. Third, by downloading RStudio (and R along with it) and installing the needed packages directly from within that GUI by a simple point-and-click process, everything needed to run M
                            <span class="s4">4</span>
                            can be obtained very easily by one-stop (free) shopping with very little or no system configuration on the user’s part, further improving access. And fourth, it’s one of the most widely used scientific computing languages, and therefore likely to be familiar to a relatively wide range of users.
                        </p>
                        <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">Basically, it came down to a question of some specifical technical requirements plus general accessibility.</p>
                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            The M
                            <span class="s4">4</span>
                            development process began in early 2017 with a partnership between NRCS and White Rabbit R&amp;D LLC facilitated through Elyon International Inc. Over the next three years, through a collaborative and consultative process, the existing PCR-based VIPER system at NRCS was evaluated, alternative approaches were explored under the constraint that only data-driven methods were to be considered, the attractiveness of developing a new system was identified along with its desired characteristics, and an initial research prototype was jointly developed with some preliminary testing, documentation, and external vetting through the engineering journal peer-review process. Work on M
                            <span class="s4">4</span>
                            since 2019 has been undertaken exclusively by NRCS and has included extensive retrospective testing, live operational testing for selected test sites during the 2020 and 2021 water supply forecast seasons, some further code debugging and refinement, and extensive documentation with additional external vetting through scientific journal peer-review processes, leading to the more refined and better-proven prototype archived here.
                        </p>
                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            Considerable effort has also been undertaken by NRCS in partnership with a software development team at Colorado State University to build Persephone, a prototype production platform for large-scale operational deployment of M
                            <span class="s4">4</span>
                            . The existing M
                            <span class="s4">4</span>
                            code sits inside the Persephone wrapper. Persephone adopts a Software-as-a-Service (SaaS) framework residing on a private cloud. This enables increased levels of parallelization, multi-user server-based access for an operational forecast team, and variety of enhanced user-focused practical capabilities: GUI, IT/database linkages, interactive capabilities around graphics, mapping and geospatial analysis, features engineering, data pre-processing and forecast distribution post-processing, and job-queuing, for example. A prototype of Persephone-M
                            <span class="s4">4</span>
                            became operational at NRCS on a limited testing and forecast process-augmentation basis during the 2023 water supply forecast season. Note that this GitHub distribution of the M
                            <span class="s4">4</span>
                            prototype does not include Persephone.
                        </p>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="3.6">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">
                            General workflow of M
                            <span class="s7">4</span>
                        </h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            There are two main steps in the M
                            <span class="s4">4</span>
                            WSF modeling process – building a model, and then running it in forecast operations – which closely mirror the steps in other hydrologic forecast models.
                        </p>
                        <ol id="l8">
                            <li data-list-text="3.6.1">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">The build step</p>
                                <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Training is the first step. It amounts to building a M
                                    <span class="s4">4</span>
                                    model suite, including data selection, setting some run control parameters, fitting the various constituent modeling systems within M
                                    <span class="s4">4</span>
                                    , and saving the outcomes. It can loosely be viewed as a geophysical model inversion process, in which model parameters are estimated from observational data. Training is akin to fitting a linear regression model in statistical WSF, or calibrating model parameters in a process-based model used for WSF.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    In general, and in brief, training involves the following processes within M
                                    <span class="s4">4</span>
                                    (see also foregoing sections in this manual and, in particular, the Appendices). As noted above, there are currently six semi- independent models in M
                                    <span class="s4">4</span>
                                    , each built around a different regression-type supervised learner. These are a monotone artificial neural network, a monotone composite quantile regression neural network, a
                                </p>
                                <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    support vector machine (more specifically, support vector regression), random forests, linear quantile regression, and a standard linear regression as in conventional PCR. A sample of input variables is drawn from a pool of candidates, a principal component analysis (PCA) is performed on it, and principal component time series for a certain number of PCA modes – the features, in ML jargon – are fed to one of the six constituent models as predictor variables. That model is then trained by fitting parameters, and to some extent hyperparameters through AutoML processes, using that sample dataset. The accuracy of these intermediate results is assessed. A type of global optimization approach drawn from evolutionary computing, called a genetic algorithm, then efficiently iterates through various combinations of input variables and corresponding PCA modes, ultimately selecting the most accurate combination. This is repeated independently for each of the six constituent models. (As such, each of the constituent modeling systems is a variant on the PCR approach; for example, principal component support vector machine, which we abbreviate here as PCSVM.) When completed, the results from all the models – consisting of a best estimate, and prediction intervals (uncertainty estimates) expressed as 10%, 30%, 70%, and 90% exceedance volumes – are averaged together to create the ensemble mean prediction. Because river runoff volume cannot be negative, the final training-period ensemble is tested for that condition, and if needed, the results of individual models contributing most the problem are removed one by one. In general, all of this is done internally and automatically by M
                                    <span class="s4">4</span>
                                    .
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    From a user’s perspective, the procedure only involves providing a historical dataset to M
                                    <span class="s4">4</span>
                                    for the combination of forecast point, publication date, and target period that the model is being built for, along with some instructions to M
                                    <span class="s4">4</span>
                                    about how the user would like M
                                    <span class="s4">4</span>
                                    to run for that particular problem.
                                </p>
                                <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    For the canonical use-case in statistical WSF as described above, the default instructions (this is explained more below) often don’t need to be modified in practice. In that case all the user needs to do is decide what predictor data they’d like to use – which SNOTEL snow and precipitation sites they want, maybe antecedent streamflow at the forecast point, perhaps an ENSO index thrown in for good measure, and so forth – just like in conventional statistical WSF. After the input data file containing that information is created by the user, which is an external process completed by the user in a text editor or Excel spreadsheet for instance, M
                                    <span class="s4">4</span>
                                    is then opened and run in R. For typical canonical statistical WSF use- case applications, depending on the computer and other factors, and using default M
                                    <span class="s4">4</span>
                                    parameters and settings including genetic algorithm-based feature optimization with a population size and number of generations found by experimentation to be generally suitable for such applications at a prototype level, the build run might take somewhere between 15 and 35 minutes or so. When it’s done, all important information is written to files by M
                                    <span class="s4">4</span>
                                    . This information includes the trained models.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    In general, training is performed only once, to create the trained suite of models and metadata describing them. Note that in usual operational WSF practice, however, models are updated roughly every five years or so by redeveloping them with newly acquired data. There may also be situations where the user would like to try creating a few different versions of a M
                                    <span class="s4">4</span>
                                    modeling suite, using different candidate input variables for example to test the usefulness of new kinds of predictive data, or giving the genetic algorithm in M
                                    <span class="s4">4</span>
                                    either more leeway, or less, to optimize features than in the default case.
                                </p>
                            </li>
                            <li data-list-text="3.6.2">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">The forecast step</p>
                                <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Creating operational forecasts using the previously trained model and new predictive input data is the second step in the M
                                    <span class="s4">4</span>
                                    WSF modeling process. It can loosely be viewed as a forward run of a geophysical
                                </p>
                                <p style="padding-top: 3pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    model developed through an inversion process. For example, in the initial build step, a M
                                    <span class="s4">4</span>
                                    ensemble modeling suite might first be trained to predict, on February 1, the April-July flow volume at some particular location on some particular river using, say, 30 years of historical SNOTEL data as input and observed April-July flow volumes over that same 30 year interval as the predictand. Then, in forecast operations, the user takes that trained modeling suite and runs it on February 1 of the current water year, using the present values for the SNOTEL stations, to create this year’s forecast of upcoming April- July flow volume. The user will do the same thing next water year, using the same old trained model and new values for the SNOTEL stations, when their February 1 forecast date rolls around again, and so forth. This is exactly like taking an existing fitted linear regression model in statistical WSF, or an existing calibrated process-simulation model in ESP-based WSF, and running it using new forcing data in order to issue a WSF for the current forecast season.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    From a user’s perspective, workflows for such forecast runs are in some sense much simpler than for training: the finished model and metadata around it, along with up-to-date data for the predictors, are all that’s needed. However, the user does need to set some slightly persnickety values in a run-control file; the user simply pulls that information from the output files of the build run (if the genetic algorithm was used for M
                                    <span class="s4">4</span>
                                    feature optimization during the build phase) or, even more simply, repeats information they provided to M
                                    <span class="s4">4</span>
                                    during the build run (if the user switched off genetic algorithm-based feature optimization during the build phase, which they have to option to do, as described below).
                                </p>
                                <p style="text-indent: 0pt;text-align: left;">
                                    <br>
                                </p>
                            </li>
                        </ol>
                    </li>
                </ol>
            </li>
            <li data-list-text="4">
                <h3 style="padding-left: 44pt;text-indent: -36pt;text-align: left;">
                    How to use M
                    <span class="s3">4</span>
                </h3>
                <ol id="l9">
                    <li data-list-text="4.1">
                        <h4 style="padding-top: 9pt;padding-left: 43pt;text-indent: -36pt;text-align: left;">
                            M
                            <span class="s7">4</span>
                            installation
                        </h4>
                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            The M
                            <span class="s4">4</span>
                            prediction analytics engine consists of 14 R script files, described in Section 4.2.1 below. These are a main program (MMPE-Main_MkII.R) plus 13 externally defined functions which perform principal component analysis, neural network modeling, genetic algorithm-based feature optimization, and so forth. The “MMPE” acronym refers to a previous name (multi-model prediction engine) given to an earlier development version of the M
                            <span class="s4">4</span>
                            metasystem.
                        </p>
                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            Installation is straightforward and essentially the same as for more-or-less any R-based software, amounting to downloading the 14 script files and installing the required packages. M
                            <span class="s4">4</span>
                            is run by executing the main program, MMPE-Main_MkII.R.
                        </p>
                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            M
                            <span class="s4">4</span>
                            was developed, tested, and run in the RStudio freeware development environment on a Windows PC, and the following, more detailed, installation instructions are tailored to that case:
                        </p>
                        <ol id="l10">
                            <li data-list-text="1.">
                                <p style="padding-top: 7pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">Download the 14 R script files (which are simply text files with a .R extension) from the GitHub repo. To keep things tidy, you&#39;ll presumably want to create a new subdirectory to place them in; you can name it whatever you like. All the model, input, and output files will be in the same folder. It’s good practice – especially when using RStudio, which can be less stable than running R in its basic console – to place this folder high up in the directory structure, to run it on a local hard drive rather than a cloud, and to use reasonably short folder names; a good choice could be something like C:\ThisWhereMyM4RunsAre\.</p>
                            </li>
                            <li data-list-text="2.">
                                <p style="padding-top: 2pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    Download the 3 regular text files (i.e., with a .txt extension) to the same location, which are input files. MMPE_RunControlFile.txt is where the user specifies run options. The other text files are input data files (the sample input file provided here is for the Yellowstone River). As noted above, M
                                    <span class="s4">4</span>
                                    has two basic modes. In inverse or model-building mode, the AIs are trained, optimized, saved, etc., using the calibration-period observational data in MMPEInputData_ModelBuildingMode.txt. In forward or operational forecasting mode, the completed models are read in from files previously created by M
                                    <span class="s4">4</span>
                                    during the build phase, and are then ran with a new sample of the input data, contained in MMPEInputData_ForecastingMode.txt.
                                </p>
                            </li>
                            <li data-list-text="3.">
                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">In Windows file explorer, navigate to this new subdirectory and double-click on the main program file, MMPEMain_MkII.R, to open the prediction engine in RStudio. (Alternatively, open RStudio, then from the main menu bar at the top select File -&gt; Open File and choose MMPE- Main_MkII.R.)</p>
                            </li>
                            <li data-list-text="4.">
                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    You&#39;ll have to install the non-default R packages used by the M
                                    <span class="s4">4</span>
                                    main program and function files. This is done in RStudio by going to its bottom right window, selecting Packages -&gt; Install and finding the ones you want by typing their names into the bar that pops up. The full list of additional packages needed is:
                                </p>
                                <p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Akima</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">forecast</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">qrnn</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">e1071</p>
                                <p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">randomForest</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">monmlp</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">genalg</p>
                                <p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">stringr</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">doParallel</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">foreach</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">quantreg</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">quantregGrowth</p>
								<p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">matrixStats</p>
                            </li>
                            <li data-list-text="5.">
                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                    The user-selected options in MMPE_RunControlFile.txt are discussed below. However, one of the parameters the user sets there counts as an installation step. It&#39;s the number of cores in the computer&#39;s CPU, num_cores, which is used for parallelization of the time-intensive cross- validation calculations in the two neural network methods included among the six supervised learning systems in M
                                    <span class="s4">4</span>
                                    . On a Windows PC, you can find the number of cores your processor has by pressing Ctrl + Shift + Esc to open Task Manager; then select the Performance tab to see how many cores (not logical processors!) there are on your machine.
                                </p>
                            </li>
                        </ol>
                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            Details on how to run M
                            <span class="s4">4</span>
                            from within the RStudio environment are given in Section 4.3 below. As noted above, it is of course possible to run M
                            <span class="s4">4</span>
                            directly in R as well, which is less user-friendly but can have advantages like avoiding graphics-related crashes that sometimes occur in RStudio. If a crash occurs and cryptic errors are recorded in the log file (see below), a simple hack that can sometimes work is to just resize the plot (lower right) window in RStudio to make it larger and rerun. It also seems to be good
                        </p>
                        <p style="padding-top: 3pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            practice to shut down and restart RStudio after every few M
                            <span class="s4">4</span>
                            runs. Some pointers around directory placement and names that were provided above can help too.
                        </p>
                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">
                            Note that M
                            <span class="s4">4</span>
                            has thus far only been tested on a Windows PC.
                        </p>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="4.2">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">File inventory and descriptions</h4>
                        <ol id="l11">
                            <li data-list-text="4.2.1">
                                <p class="s5" style="padding-top: 9pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                                    M
                                    <span class="s6">4</span>
                                    code files
                                </p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The R files that collectively make up M
                                    <span class="s4">4</span>
                                    are as follows. All are invoked in M
                                    <span class="s4">4</span>
                                    runs during the model-build phase; in forecast runs, only the main program and the three ensemble-creation utility functions are used:
                                </p>
                                <ul id="l12">
                                    <li data-list-text="">
                                        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            <i>MMPE-Main_MkII.R</i>
                                            The M
                                            <span class="s4">4</span>
                                            main program. Clears workspace, reads in data, declares external functions, tracks errors, creates (in build mode) the six modeling systems, including ensemble creation and other tasks, along with graphics and output files or runs (in forecast mode) the six existing modeling systems using output files created during the build phase.
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCA-Module_MkII.R
                                            <span class="p">Function performing principal component analysis on the predictor variates provided to it; this may be all the predictors in the input data file, MMPEInputData_ModelBuildingMode.txt (described in detail below), if the user specifies in MMPE_RunControlFile.txt (also described in the detail below) that genetic algorithm-based feature optimization is not to be used, or a subset of the predictors in that input data file if GA- based feature optimization is used (see Sections 4.2.2 and 4.3 below). PCA is performed on the correlation, not covariance, matrix.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;text-align: left;">
                                            PCA-Graphics-Module_MkII.R
                                            <span class="p">Function creating some PCA-related plots.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            GA-PredictorSelection-Module_MkII.R
                                            <span class="p">Function performing genetic algorithm-based feature optimization, if that option is selected by user in the input file, MMPE_RunControlFile.txt; includes some automatic hyperparameter-setting capabilities around the monotone artificial neural network and monotone composite quantile regression neural network algorithms; cost function is root mean square error. Note that the cost is given a penalty if the trial feature set provided by the genetic algorithm in a given iteration includes fewer than MinNumVars of the input predictor variables provided by the user in the building-phase input data file, MMPEInputData_ModelBuildingMode.txt, where MinNumVars is settable by the user in MMPE_RunControlFile.txt but must be &gt; 1 and &lt; number of predictor variates provided by the user in MMPEInputData_ModelBuildingMode.txt and should in general be left at 2. The effect is to ensure that the genetic algorithm is guided away from cases where only a single predictor variable is retained, which is inconsistent with the canonical use-case for statistical WSF in the western US (generally framed around several multicollinear predictor variables). A single-</span>
                                        </p>
                                        <p style="padding-top: 3pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            predictor application of M
                                            <span class="s4">4</span>
                                            is possible for unusual circumstances but does not use genetic algorithm feature optimization (see use instructions below in Section 4.3).
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCR-Module_MkII.R
                                            <span class="p">Function performing linear regression model-fitting along with cross- validated performance evaluation and estimation of both stationary Gaussian, and heteroscedastic non-Gaussian, prediction intervals around the best-fit PCR estimates; the latter uses a method based on a Box-Cox transform on the prediction residuals, including estimation of the best-fit Box-Cox transform parameters.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCSVM-Module_MkII.R
                                            <span class="p">As in PCR-Module_MkII.R but for the support vector machine; additionally includes some AutoML capabilities for identifying optimal SVM hyperpameters in accordance with options set by user in in the input file, MMPE_RunControlFile.txt.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;text-align: left;">
                                            PCRF-Module_MkII.R
                                            <span class="p">As in PCR-Module_MkII.R but for random forests.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCANN-Module_MkII.R
                                            <span class="p">As in PCR-Module_MkII.R but for a monotone artificial neural network; additionally includes setting some hyperparameters in accordance with values specified by user in MMPE_RunControlFile.txt and/or delivered to it by AutoML algorithms in MMPE-Main_MkII.R and/or GA-PredictionSelection-Module_MkII.R; uses (if option enabled in MMPE_RunControlFile.txt) light parallelization of embarrassingly parallel ANN cross-validation task across process cores.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCQR-Module_MkII.R
                                            <span class="p">As in PCR-Module_MkII.R but using quantile regression with enforcement of a quantile non-crossing constraint that can be crucial for small sample sizes; expectation value is taken to be the median; heteroscedastic non-Gaussian prediction intervals around the best-fit QR predictions are estimated using the linear quantile-regression capabilities inherent to the method; due to some quirks with one of the R packages used, while a .Rdata file is produced, the primary information about the trained model is written to a text file instead, unlike the other six modeling systems.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            PCMCQRNN_MkII.R
                                            <span class="p">As in PCANN-Module_MkII.R, but for the monotone composite quantile regression neural network; expectation value is taken to be the median; heteroscedastic non- Gaussian prediction intervals around the best-fit predictions are estimated using the nonlinear quantile-regression capabilities inherent to the method.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p class="s2" style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            InitializeEnsemble-Module_MkII.R, AppendEnsemble-Module_MkII.R, FinalizeEnsemble- Module_MkII.R
                                            <span class="p">Utility functions used in creating multi-model ensemble in accordance with user-specified preferences set in MMPE_RunControlFile.txt input file.</span>
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            <i>Diagnostics-Module_MkII.R</i>
                                            Creates visual diagnostics and statistical measures describing model fit for each of the six constituent modeling systems and the multi-model ensemble mean forecast distribution. Some of these results are written to file, whereas others are graphically depicted in windows that open in R during an M
                                            <span class="s4">4</span>
                                            run.
                                        </p>
                                    </li>
                                </ul>
                                <p style="padding-top: 3pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Note that for standard- or intermediate-level use of M
                                    <span class="s4">4</span>
                                    (see below) no edits are required to any of these files. In other words, it’s strongly suggested you don’t mess with them unless you really know what you’re doing!
                                </p>
                            </li>
                            <li data-list-text="4.2.2">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Input files</p>
                                <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">
                                    M
                                    <span class="s4">4</span>
                                    has only three input files, and only two of them are used for a given run of the program:
                                </p>
                                <ul id="l13">
                                    <li data-list-text="">
                                        <p style="padding-top: 9pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            <i>MMPEInputData_ModelBuildingMode.txt</i>
                                            The input data file for the training phase of M
                                            <span class="s4">4</span>
                                            application. These are the historical data used to fit the M
                                            <span class="s4">4</span>
                                            modeling suite. The file is only required during the build step. Don’t change the name of the file.
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            The file contains a (
                                            <i>N</i>
                                            +1) row by (
                                            <i>M</i>
                                            +2) column matrix, where
                                            <i>N</i>
                                            is the number of samples and
                                            <i>M</i>
                                            is the number of predictors. The first column is, in the canonical use-case for western US statistical WSF, the year of the observation but could conceivably be a different sort of numerical sample index; the second column gives the observed flow volume in that year, i.e., the predictand; and subsequent columns give the observed values of various predictor variables in that year, which in the canonical use-case will typically be quantities like SWE, accumulated precipitation, and so forth (see previous sections). The top row is a header, providing information about each of the columns; avoid using ambiguous wildcard characters or spaces in the title for a given column. Columns are tab-separated.
                                        </p>
                                        <p style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            If the genetic algorithm in M
                                            <span class="s4">4</span>
                                            is employed for feature optimization, these predictor variables collectively constitute a pool of candidates, and the GA will pick the ones that work best (see definition of ‘best’ in section 4.2.1 or the Appendices) for a given model. This is done independently for each of the six constituent modeling systems, as each of the corresponding supervised regression-type models have their own strengths and weaknesses and may work best for a different subset of the candidate predictors. The canonical western US statistical WSF use-case is built around several multicollinear predictors. If feature optimization is used, at least three, and in normal WSF use typically somewhere around a half-dozen to two-dozen, candidate input variables would be provided in this file.
                                        </p>
                                        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            If on the other hand the genetic algorithm is not used in M
                                            <span class="s4">4</span>
                                            , all the input variables provided in this file will be used in all the models. In this case, only one predictor variable is necessary (though for canonical western US statistical WSF applications more should generally be provided; see above).
                                        </p>
                                        <p style="padding-top: 8pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">Here’s a complete example of the contents of MMPEInputData_ModelBuildMode.txt, using data from 1986 through 2015:</p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <table style="border-collapse:collapse;margin-left:77.3404pt" cellspacing="0">
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">Year</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">Obs</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">LM_P</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">LM_SWE</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 4pt;text-indent: 0pt;line-height: 8pt;text-align: center;">Sig_P</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">Sig_SWE</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">Sil_P</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">Sil_SWE</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1986</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">33.036</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.3</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">13.8</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">16.6</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.3</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1987</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">50.664</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.2</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.1</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">15.8</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.9</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">19.8</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.5</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1988</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">34.106</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.2</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.2</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">11.5</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.7</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">17.1</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.8</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1989</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.24</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.7</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 14pt;text-indent: 0pt;line-height: 8pt;text-align: center;">7.2</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.8</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.4</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1990</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">15.647</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.4</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.1</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 14pt;text-indent: 0pt;line-height: 8pt;text-align: center;">8.3</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.3</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.8</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.2</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1991</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">97.577</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.8</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.6</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">16.4</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.9</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">20.2</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.8</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1992</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">116.254</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.2</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.8</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">14.7</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.3</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">20</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.5</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1993</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">78.318</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.7</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 3pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">28.4</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.8</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">29.2</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">20.4</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1994</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">18.621</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.5</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.5</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 2pt;padding-right: 14pt;text-indent: 0pt;line-height: 8pt;text-align: center;">8.3</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.5</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.2</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.2</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1995</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">39.69</p>
                                                </td>
                                                <td style="width:29pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.5</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.1</p>
                                                </td>
                                                <td style="width:33pt">
                                                    <p class="s8" style="padding-left: 2pt;padding-right: 9pt;text-indent: 0pt;line-height: 8pt;text-align: center;">23.6</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.3</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 1pt;text-indent: 0pt;line-height: 8pt;text-align: left;">21.6</p>
                                                </td>
                                                <td style="width:42pt">
                                                    <p class="s8" style="padding-left: 6pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.1</p>
                                                </td>
                                            </tr>
                                        </table>
                                        <table style="border-collapse:collapse;margin-left:77.2446pt" cellspacing="0">
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1996</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.884</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.4</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 11pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.5</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.5</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.5</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.1</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1997</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">43.454</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.7</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.8</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 11pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.8</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.9</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">15.2</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.4</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1998</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">84.973</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.3</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 11pt;text-indent: 0pt;line-height: 8pt;text-align: left;">14.8</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.4</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">18.3</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.7</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1999</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.171</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.4</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.4</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2000</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.007</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.1</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.9</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.2</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.1</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2001</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">34.079</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.2</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.4</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">14.3</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.7</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">22.8</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.8</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2002</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.822</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.5</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.1</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.4</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.6</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.1</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2003</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">19.799</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.1</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.2</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.4</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.2</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.8</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2004</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">46.132</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.8</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.4</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.2</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.5</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.3</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.1</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2005</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">54.77</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">10.2</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.3</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">21.5</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.2</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">22.5</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.5</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2006</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.223</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.3</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.1</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.4</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.6</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2007</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">33.017</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.3</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.8</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.3</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.6</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.6</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7.5</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2008</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">30.537</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.2</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.6</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.3</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">19.1</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.4</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2009</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">14.418</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.3</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.9</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.1</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.2</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2010</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">93.842</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.5</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">15.5</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.4</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">18.7</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">16.4</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2011</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.288</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.8</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.7</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.5</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.9</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2012</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">16.623</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.3</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">12.3</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.9</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">11.5</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">7</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2013</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">19.035</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.7</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0.4</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.2</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.5</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2014</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">13.127</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">3.6</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2.6</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">5.2</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">1.3</p>
                                                </td>
                                            </tr>
                                            <tr style="height:10pt">
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 2pt;text-indent: 0pt;line-height: 8pt;text-align: left;">2015</p>
                                                </td>
                                                <td style="width:41pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">22.494</p>
                                                </td>
                                                <td style="width:31pt">
                                                    <p class="s8" style="padding-left: 3pt;text-indent: 0pt;line-height: 8pt;text-align: left;">6.5</p>
                                                </td>
                                                <td style="width:34pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s8" style="padding-left: 10pt;text-indent: 0pt;line-height: 8pt;text-align: left;">8.1</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">0</p>
                                                </td>
                                                <td style="width:36pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">9.8</p>
                                                </td>
                                                <td style="width:30pt">
                                                    <p class="s8" style="padding-left: 8pt;text-indent: 0pt;line-height: 8pt;text-align: left;">4.4</p>
                                                </td>
                                            </tr>
                                        </table>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <p style="padding-top: 3pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            This is the predictand and forcing data to train a M
                                            <span class="s4">4</span>
                                            modeling suite that forecasts, on March 1, the total March-through-May flow volume of the Gila River near Gila in New Mexico. Thirty years of historical data are used in this example. The first column is the year of observation, the second column is the observed March 1-May 31 Gila flow volume at this location over those years, the third and fourth columns give the water-year-to-date (October 1 through February 28) accumulated precipitation and March 1 SWE at the Lookout Mountain SNOTEL site, the fifth and sixth columns do the same for the Signal Peak SNOTEL site, and the seventh and eighth columns do the same for the Silver Creek Divide SNOTEL site. If the genetic algorithm is invoked by the user, M
                                            <span class="s4">4</span>
                                            will pick, independently for each of its constituent modeling systems, which of these six predictors to keep. If the genetic algorithm isn’t used, all six predictors in this example will be used in each of the constituent modeling systems.
                                        </p>
                                        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Here are a few tips on creating this input dataset. M
                                            <span class="s4">4</span>
                                            isn’t a magical exception to the garbage- in, garbage-out rule. One example of something to avoid is a time series predictor (i.e., a column in the data matrix show above) with no, or almost no, variability (e.g., a SWE value from some SNOTEL site that is 0 or near-0 in all, or almost all, years for the Julian day of the forecast, which can happen for, say, a low-elevation SNOTEL site in a late-season forecast after significant snowmelt has occurred). Such a constant-valued dataset in general has no use for making predictions. More generally, the user has to employ some professional judgement – potentially based on prior experience with the watershed, and exploratory statistical or graphical analysis – to choose predictors for the input data matrix that are likely to have some potential to offer predictive value. Also use sufficient sample sizes. M
                                            <span class="s4">4</span>
                                            ’s use of PCA-based dimensionality reduction, compact ML architectures, and a priori monotonicity and nonnegativity constraints in default applications collectively result in a comparatively small number of parameters having to be trained and, in effect, only over restricted ranges (see again background above and in the appendices). This has a few advantages, one of which is that it allows the use of what would normally be considered, in the AI community, tiny datasets – like the one shown above, containing only 30 samples. Testing has suggested it may have capacity to work effectively in canonical western US statistical WSF use-cases as described above with as little as 20 samples. But 10 samples, for example, is insufficient to fit and test any statistical WSF model to be used in operational applications with much confidence, including even a simple linear regression, and
                                        </p>
                                        <p style="padding-top: 2pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            it’s blatantly inadequate for a ML approach. Mistakes like these will result at best in poor models and at worst in a crash of one of the constituent machine learning algorithms. More broadly, recall that M
                                            <span class="s4">4</span>
                                            was explicitly and specifically designed to leverage the professional expertise of the hydrologic modeler during the process of features engineering – in particular, when creating the input data pool – as a pragmatic means toward achieving theory-guided and explainable machine learning.
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            <i>MMPEInputData_ForecastingMode.txt</i>
                                            The input data file for the forecasting phase of M
                                            <span class="s4">4</span>
                                            application. It must be exactly identical to the MMPEInputData_ModelBuildingMode.txt file used to build the M
                                            <span class="s4">4</span>
                                            model suite that’s now being run in forecast mode, except:
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <ul id="l14">
                                            <li data-list-text="o">
                                                <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">There aren’t any columns for the year or for the observed flow volume; and</p>
                                            </li>
                                            <li data-list-text="o">
                                                <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">Below the header, there’s only a single row of data, which is the present value for the various candidate input predictors included in the build phase, arranged left to right in exactly the same order as they were in MMPEInputData_ModelBuildingMode.txt.</p>
                                                <p style="padding-top: 8pt;padding-left: 43pt;text-indent: 0pt;text-align: left;">Here is a complete example for the Gila River case considered immediately above:</p>
                                                <p class="s9" style="padding-top: 9pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">LM_P  LM_SWE Sig_P Sig_SWE Sil_P Sil_SWE</p>
                                                <p class="s9" style="padding-left: 79pt;text-indent: 0pt;text-align: left;">3.6  0   2.6  0   5.2  1.3</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                                <p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                    Given that the model was built on 1986-2015 data, this might be the observed values during forecast operations on, say, March 1, 2016, or March 1, 2017, and so forth. Note that the predictors, going from left to right (Lookout Mountain SNOTEL accumulated year-to-date precipitation, Lookout Mountain SNOTEL March 1 SWE, etc.) are in exactly the same order as in the training data file (see above). This file is only required when M
                                                    <span class="s4">4</span>
                                                    is being run in forecast mode. Don’t change the name of the file.
                                                </p>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="">
                                        <p style="padding-top: 7pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            <i>MMPE_RunControlFile.txt</i>
                                            The user specifies here, by setting various parameters or flags, how they would like M
                                            <span class="s4">4</span>
                                            to run. This includes whether a build or forecast run is desired; for a build run, some controls over how that is approached; and for a forecast run, some outcomes from the build run and a couple options. This file is required for every M
                                            <span class="s4">4</span>
                                            run. The MMPE_RunControlFile.txt example provided in the M
                                            <span class="s4">4</span>
                                            installation package provides:
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <ul id="l15">
                                            <li data-list-text="o">
                                                <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                    Default values for every parameter or flag. Testing has suggested that these settings work well for the canonical use-case for statistical WSF in the western US. Such standard WSF use should generally not require diverging from these default values (see Section 4.3 below). There are two exceptions: num_cores, which is a parameter set by the user as part of the M
                                                    <span class="s4">4</span>
                                                    installation (see step 5 in Section 4.1 above), and a few settings in the bottom, forecast-mode portion of the file that need to be set by the user on the basis of the results from the build-mode M
                                                    <span class="s4">4</span>
                                                    run.
                                                </p>
                                            </li>
                                            <li data-list-text="o">
                                                <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                    Annotation, on each line, of the variable name used within the M
                                                    <span class="s4">4</span>
                                                    code for that parameter or flag, followed by a brief description and some guidance on its allowable values and format.
                                                </p>
                                            </li>
                                        </ul>
                                    </li>
                                </ul>
                                <p style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">The following is drawn from the example file provided in the distribution. (Open the file and compare to the descriptions provided below.) The file is divided into three sections.</p>
                                <p style="padding-top: 2pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">The first section simply specifies whether a build or forecast run is desired, and whether the user wishes to switch on or off a log file:</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;"># BASIC PARAMETERS: CONTROL WHETHER INVERSE OR FORWARD RUN, AND WHETHER MESSAGES ARE SENT TO AN EXTERNAL LOG FILE</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;BUILD&quot;
                                    <span class="p">(variable name: RunTypeFlag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Write in “BUILD” if a build run is desired, or “FORECAST” if a forecast run is desired. Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: errorlog_flag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">Do you want a log file to be created? Enter “Y” or “N”, again all caps, don’t drop the quotation marks. Generally not much reason to use anything other than “Y” but the user is given the choice.</p>
                                <p style="padding-top: 8pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    In the second section, the user provides information about how they would like a build run to proceed. Correctly formatted values must always be provided; however, M
                                    <span class="s4">4</span>
                                    ignores the values of these parameter and flags if RunTypeFlag = “FORECAST” (see above):
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># RUN PARAMETERS FOR MODEL DEVELOPMENT (Used only if RunTypeFlag</p>
                                <p class="s10" style="padding-left: 80pt;text-indent: 0pt;text-align: left;">= &quot;BUILD&quot;)</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># CONTROL FEATURE CREATION &amp; SELECTION:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: GeneticAlgorithmFlag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Use the genetic algorithm in M
                                    <span class="s11">4</span>
                                    for optimal predictor selection? If “Y”, code finds optimal combination of input variables, and which PCA modes to retain, separately for each of the modeling technologies, and fits each of the models accordingly. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: justify;">
                                    2
                                    <span class="p">(variable name: MaxModes)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: justify;">If GeneticAlgorithmFlag = &quot;Y&quot; (ignored otherwise): specify the maximum number of PCA modes the genetic algorithm is allowed to retain as it tries to optimize feature selection, counting upward from the leading mode, which is mode 1.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">The leading mode is always retained, so the minimum allowable value of MaxModes is 1. Note that if MaxModes is set to its minimum allowable value of 1, i.e., if only the leading PCA mode is retained, then the genetic algorithm only optimizes which predictors in the candidate pool to retain.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If MaxModes is set to 2, for example, the genetic algorithm may retain PCA modes 1 and 2; or it may only retain the leading mode, depending on whether or not mode 2 helps improve prediction skill. Note that the genetic algorithm may skip modes (except for mode 1, which as noted is always retained), e.g., if MaxModes is 4 it might choose to keep modes 1, 2, and 4, skipping 3; in practice, though, this generally doesn’t happen often for the canonical WSF use-case described above in this manual.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The following is some guidance and constraints around the largest allowable value. The largest value MaxModes can take is 4, that is, the 4th PCA mode is highest available in the current version of M
                                    <span class="s11">4</span>
                                    ; this is thought to be generally far more than needed for
                                </p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">canonical western US WSF use-cases, where often only the first, and rarely more than the first and second, PCA modes are retained in operational WSF models. Additionally, because a principal component analysis of a M-variable by N-sample dataset will produce M modes, MaxModes must be also &lt; the number of predictors that the user provides in the candidate predictor pool of MMPEInputData_ModelBuildMode.txt. It should also be recalled that a goal of PCA in applications like the canonical statistical WSF use-case in the western US is data compression/dimensionality reduction from a multicollinear data pool, so that as a matter of best-practice, MaxModes should in general be set &lt;&lt; M.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: DisplayFlag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If GeneticAlgorithmFlag = &quot;Y&quot; (ignored otherwise), display to screen the progress of genetic algorithm? (note: progress written to file irrespective of this setting). Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    2
                                    <span class="p">(variable name: MinNumVars)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If GeneticAlgorithmFlag = &quot;Y&quot; (ignored otherwise): specify the minimum allowable number of input variables to retain. Discussion of this parameter and how it’s used was provided in the description of GA-PredictorSelection-Module_MkII.R in Section 4.2.1. It cannot be lower than 2, and obviously it can’t be higher than the number of candidate predictors provided by the user in the input data file, MMPEInputData_ModelBuildingMode.txt. In general, the default value of 2 can be used.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    25
                                    <span class="p">(variable name: GAPopSize)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If GeneticAlgorithmFlag = &quot;Y&quot; (ignored otherwise): specify the population size to be used in the genetic algorithm. The larger the population, the better (roughly speaking) the result, up to a point, but the longer the build run. GAPopSize = 25 with GANumGens = 10 (see below) is a good compromise for canonical WSF use-cases of the type described above in this manual and is a recommended default. If in a rush, GAPopSize = 15 with GANumGens = 7 generally gives usable results with a shorter run time. If the size of the candidate predictor pool is larger, or if build-phase run time is not an issue, it may be a good idea to run a larger genetic algorithm problem to help ensure that the global best estimate is obtained. GAPopSize = 100 with GANumGens = 15, for example, can take hours to turn but will more reliably obtain a truly optimal feature selection for a canonical WSF use-case scenario, though the improvement will often be modest relative to the default, and there’s a point of diminishing returns with increasing GAPopSize and GANumGens.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    10
                                    <span class="p">(variable name: GANumGens)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If GeneticAlgorithmFlag = &quot;Y&quot; (ignored otherwise): specify the number of generations used in the genetic algorithm. See description of GAPopSize above.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1,2
                                    <span class="p">(variable name: ManualPCSelection)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If GeneticAlgorithmFlag = &quot;N&quot; (ignored otherwise): specify which PCA modes to fit the models to. 1 = leading mode, 2 = second mode, etc., 3 = third mode, etc.; use comma between modes if more than one retained; the same modes will be used for all modeling</p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">techniques; all input variables in the input data file will be used; modes may be skipped, except for the leading mode; the 4th PCA mode is highest allowable; recall that a PCA returns only as many modes as there are input variables to it, and that a goal of PCA in an application like this is data compression/dimensionality reduction, so in general it must be no larger than and generally should be &lt;&lt; M, the number of predictor variables provided by the user in MMPEInputData_ModelBuildingMode.txt.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># CONTROL ENSEMBLE GENERATION:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: AutoEnsembleFlag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">Use automatic ensemble generation? If &quot;Y&quot;, the code starts with default ensemble members, automatically checks ensemble output for non-physical (negative) values, and adjusts member composition accordingly.</p>
                                <p class="s2" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">The 6 default ensemble members are linear regression with post hoc Box-Cox transform space-based prediction intervals (denoted PCR_BC below); quantile regression (PCQR); a monotone artificial neural network with Box-Cox transform space-based prediction intervals (PCANN_BC); a monotone composite quantile regression neural network (PCMCQRNN); random forests with Box-Cox transform space-based prediction intervals (PCRF_BC); and a support vector machine with Box-Cox transform space-based prediction intervals (PCSVM_BC).</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If instead set to “N”, the user specifies which models to include in the ensemble mean forecast distribution using the next 10 run control parameters described immediately below. This capability is rarely used, and AutoEnsembleFlag should generally be left at the default, “Y”. In that case, the values used for the next 10 run control parameters below don’t matter, though some value, in the correct format, has to be provided.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">However, it can be useful to provide the user with some flexibility around which models, and which variants of which models, to include in the ensemble mean forecast distribution.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    For example, PCR and PCR_BC (see below) are the same linear regression model, producing the same best estimate, but one (PCR) uses conventional (stationary Gaussian) prediction intervals around the best-estimate, much like the current NRCS PCR model, and the other (PCR_BC) uses prediction intervals calculated in a similar fashion but in Box-Cox transform space. The latter is in general superior because it seamlessly and automatically accommodates non-Gaussian and heteroscedastic error distributions yet can also adequately capture Gaussian and homoscedastic conditions when appropriate. However, once in a blue moon, PCR_BC produces unrealistically wide prediction intervals, a quirk that will be addressed in some future version of M
                                    <span class="s11">4</span>
                                    , and in such cases the user may prefer to use PCR instead in the final ensemble mean forecast distribution. Similar considerations apply for PCANN vs. PCANN_BC, PCRF vs. PCRF_BC, and PCSVM vs. PCSV_BC. Note that in this case, the default values for the next 10 run control parameters as provided in the descriptions below and in the M
                                    <span class="s11">4</span>
                                    standard distribution are set in such a way that, if the user wishes to avoid the Box-Cox transform space-based prediction intervals altogether, the user merely has to switch
                                </p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">AutoEnsembleFlag from “Y” to “N”. Recall, though, that this also turns off the negativity-check and ensemble-pruning algorithm.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    (Note also that in very rare cases, the post hoc Box-Cox transform space-based prediction intervals are so wide they include infinity, which would cause subsequent visualization steps in M
                                    <span class="s11">4</span>
                                    , such as diagnostics plots, to crash. To avoid this, an algorithm tests for Inf values in any of the prediction intervals for PCR_BC, and if found, all the PCR_BC prediction intervals are replaced by those from PCR. In this case, the outputs for PCR and PCR_BC are identical and represent PCR only. When this happens, it is immediately obvious on inspection of the output files; additionally, one of the output files (see detailed list of output files provided in a subsequent section) explicitly tracks whether this has happened, so there is no room for confusion as to what has been produced. The same applies for the other three models in M
                                    <span class="s11">4</span>
                                    that are available with both conventional prediction intervals and Box-Cox transform space-based intervals: the monotone artificial neural network, support vector machine, and random forests. None of this applies to the two (nonlinear neural network, and linear) quantile regression methods, which intrinsically generate non-Gaussian heteroscedastic predictions intervals.)
                                </p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">As another example of an unusual case in which the user might wish to switch AutoEnsembleFlag to N, in an application that is known a priori to be strongly nonlinear, the user may wish to exclude linear regression (PCR) and linear quantile regression (PCQR) from the ensemble.</p>
                                <p class="s2" style="padding-top: 8pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Note, though, that all training runs of M
                                    <span class="s11">4</span>
                                    produce all 10 (the 6 defaults, plus non-Box Cox prediction interval versions of PCR, PCRF, PCANN, and PCSVM) sets of individual model outputs and write them to files, so the user can just manually create their own multi-model ensemble mean with a fully custom ensemble composition post-hoc outside of M
                                    <span class="s11">4</span>
                                    . This may be more practical than setting AutoEnsembleFlag to “N” and specifying which ensemble members to retain using the following flags, because in many cases you might not know until after you’ve completed the entire M
                                    <span class="s11">4</span>
                                    build run, and evaluated its results, whether you want to change the ensemble composition. In general, however, one would generally want to avoid using, for example, both PCANN and PCANN_BC, as they differ only in their prediction intervals, and this therefore doesn’t add another independent best-estimate to the multi-model ensemble. Note also that the diagnostics M
                                    <span class="s11">4</span>
                                    produces for the final multi-model ensemble mean it generates, like statistical measures of model fit in the output file ensemble_StandardPerformanceReportingSuite.csv (see output file list below), are based on the ensemble composition it generates, not what the user might have manually created after the run finished.
                                </p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    AutoEnsembleFlag pertains to the training phase of M
                                    <span class="s11">4</span>
                                    . A forecast run of M
                                    <span class="s11">4</span>
                                    (see below) produces and writes to file a prediction from all 10 of these models as well. For details on the ensemble mean forecast distribution that forward runs generate, see below; note that if the user decided on the basis of the training run to keep a custom mix of constituent models, this may not be reflected in the forward-run ensemble mean, and
                                </p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">the user may have to construct the forward-run ensemble mean manually in accordance with what they decided the model mix should be on the basis of the training run. Recall that it is generally poor practice to use different mixes of constituent models in the forecast multi-model ensemble mean than in the training-phase multi-model ensemble mean (in part because if the user takes such an unusual step, the out-of-sample performance measures for the training-phase multi-model ensemble mean, like cross- validated RMSE, may not speak to the performance that could be expected from the different custom mix of models used in the forward run).</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Overall, the current version of M
                                    <span class="s11">4</span>
                                    is built around the six default models and the negativity-check and ensemble-pruning algorithm (i.e., AutoEnsembleFlag = Y), which testing has suggested works well for the canonical western US WSF use-case described above. Long story short: testing and experience to date applying M
                                    <span class="s11">4</span>
                                    to NRCS water supply forecast problems suggests that the user should, generally speaking, only stray away from this default if they can clearly and objectively demonstrate a good reason to, and if they are prepared to take on the additional training, responsibility, and work around building and handling custom ensemble compositions.
                                </p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">Use caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 9pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from conventional linear regression (actually principal component regression because of the PCA preprocessing) in the multi-model ensemble? The prediction intervals in this case are calculated using a simple heuristic widely used in WSF applications of linear PCR assuming Gaussian homoscedastic error distributions (see review in Section 3.2.1 and the Appendices). The best-fit model and its best-estimate predictions are identical to PCR_BC (see immediately below) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;N&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCR_BC)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from linear regression (again, actually principal component regression because of the PCA preprocessing) with Box-Cox transform-space prediction bounds in the multi-model ensemble? This is a widely used heuristic for estimating prediction intervals in WSF applications of PCR but is modified here to allow non- Gaussian and heteroscedastic error distributions, though in general it can also adequately capture Gaussian and homoscedastic distributions if appropriate. It is one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). The best-fit model and its best-estimate predictions are identical to PCR (see immediately above) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCQR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from quantile regression (actually principal component quantile regression due to the use of PCA preprocessing, hence “PCQR”) in the multi- model ensemble? Quantile regression is a variant of linear regression that calculates multiple regression lines representing quantiles of the error distribution, and permits non-Gaussian and heteroscedastic residuals. It is one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCANN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the monotone artificial neural network (actually a principal component artificial neural network because of the PCA preprocessing, hence “PCANN”) in the multi-model ensemble? The prediction intervals in this case are calculated using a simple heuristic widely used in WSF applications of linear PCR assuming Gaussian homoscedastic error distributions (see EnsembleFlag_PCR above). The best-fit model and its best-estimate predictions are identical to PCANN_BC (see immediately below) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;N&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCANN_BC)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the artificial neural network (actually a principal component artificial neural network because of the PCA preprocessing, hence “PCANN”) with Box-Cox transform-space prediction bounds in the multi-model ensemble? This is a widely used heuristic for estimating prediction intervals in WSF applications of PCR but is modified to allow non-Gaussian and heteroscedastic error distributions (see EnsembleFlag_PCR_BC above). It is one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). The best-fit model and its best-estimate predictions are identical to PCANN (see immediately above) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    “N”
                                    <span class="p">(variable name: EnsembleFlag_PCRF)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the random forests (actually a principal component random forests because of the PCA preprocessing, hence “PCRF”) in the multi-model ensemble? The prediction intervals in this case are calculated using a simple heuristic widely used in WSF applications of linear PCR assuming Gaussian homoscedastic error distributions (see EnsembleFlag_PCR above). The best-fit model and its best-estimate predictions are identical to PCRF_BC (see immediately below) but the prediction intervals</p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">are, in general, different. &quot;Y&quot; = yes, &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;N&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCRF_BC)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the random forests with Box-Cox transform-space prediction bounds in the multi-model ensemble? This is a widely used heuristic for estimating prediction intervals in WSF applications of PCR but is modified to allow non- Gaussian and heteroscedastic error distributions (see EnsembleFlag_PCR_BC above). It is one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). The best-fit model and its best-estimate predictions are identical to PCRF (see immediately above) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCMCQRNN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the monotone composite quantile regression neural network (actually principal component monotone composite quantile regression neural network due to the use of PCA preprocessing, hence “PCMCQRNN”) in the multi-model ensemble? This is an ANN variant that calculates multiple nonlinear regression lines representing quantiles of the error distribution, and permits non-Gaussian and heteroscedastic residuals. It is one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: EnsembleFlag_PCSVM)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the support vector machine (actually a principal component support vector machine because of the PCA preprocessing, hence “PCSVM”) in the multi-model ensemble? The prediction intervals in this case are calculated using a simple heuristic widely used in WSF applications of linear PCR assuming Gaussian homoscedastic error distributions (see EnsembleFlag_PCR above). The best-fit model and its best-estimate predictions are identical to PCSVM_BC (see immediately below) but the prediction intervals are, in general, different. &quot;Y&quot; = yes, &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.</p>
                                <p style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;text-align: left;">&quot;N&quot;      (variable name: EnsembleFlag_PCSVM_BC)</p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoEnsembleFlag = &quot;N&quot; (ignored otherwise, but a value is always required): for the case where the user manually specifies a la carte which models to include in the ensemble, include output from the support vector machine with Box-Cox transform- space prediction bounds in the multi-model ensemble? This is a widely used heuristic for estimating prediction intervals in WSF applications of PCR but is modified to allow non- Gaussian and heteroscedastic error distributions (see EnsembleFlag_PCR_BC above). It is</p>
                                <p class="s2" style="padding-top: 3pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    one of the default models in M
                                    <span class="s11">4</span>
                                    when automatic ensemble generation is used (see above). The best-fit model and its best-estimate predictions are identical to PCSVM (see immediately above) but the prediction intervals are, in general, different. Capital &quot;Y&quot; = yes, capital &quot;N&quot; = no. Use all caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;"># CONTROL CONFIGURATION &amp; PARALLELIZATION OF NEURAL NETWORKS (BOTH mANN AND MCQRNN):</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: AutoANNConfigFlag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Use automatic configuration selection in both the artificial neural network (ANN) and monotone composite quantile regression neural network (MCQRNN)? If &quot;Y&quot;, code starts with a basic configuration, automatically compares prediction quality to that of other models, and uses an alternative configuration if necessary and appropriate, using information theoretic and other criteria. “Configuration” in this context refers to various hyperparameters controlling network properties like neural network topology and bootstrap aggregation. The available configurations in this AutoML step were determined by extensive experimentation with canonical use-cases for statistical WSF in the western US (see above). It is recommended for most WSF users that this be left at the default value of “Y”. Use all caps and don’t drop the quotation marks.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    25
                                    <span class="p">(variable name: ANN_config1_cutoff)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoANNConfigFlag = &quot;Y&quot; (ignored otherwise, but some value is always required): this is a hyperhyperparameter in the AutoML routine for autoconfiguring the two neural networks (ANN and MCQRNN) in M
                                    <span class="s11">4</span>
                                    ; maximum allowable % performance deficit in R
                                    <span class="s11">2</span>
                                    or RMSE of default configuration relative to other (non-neural) models.
                                </p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: mANN_config_selection)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If AutoANNConfigFlag = &quot;N&quot; (ignored otherwise): if 1, use basic monotone artificial neural network configuration (1 hidden-layer neuron with no bagging), if 2 use standard backup (alternative) configuration (2 hidden-layer neurons with 20 bootstraps), if 3 use custom configuration (to be hard-coded by an experienced user within the ANN module itself). Note that because there is only one output neuron (predictand/target), and a maximum of four and typically only one or two input neurons (predictor PCA modes), the neural networks used here are, deliberately and by construction, extremely parsimonious; moreover, the monotonicity constraint generally used (see below) for applications of M
                                    <span class="s11">4</span>
                                    to canonical WSF use-cases greatly reduces the flexibility required by the neural networks. Hence, very compact hidden layers are suitable for the canonical use-case applications that are the primary focus to date of M
                                    <span class="s11">4</span>
                                    development and testing. Expert users, however, may invoke mANN_config_selection = 3 and hard-code the module implementing the ANN to use a multilayer perceptron topology just about as complex as they like.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: MCQRNN_config_selection)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If AutoANNConfigFlag = &quot;N&quot; (ignored otherwise): As above for mANN_config_selection but for the monotone composite quantile regression neural network</p>
                                <p class="s10" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: ANN_monotone_flag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Enable monotonicity constraint in the monotone artificial neural network and monotone composite quantile regression neural network? If &quot;Y&quot; then relationships between all predictors (i.e., principal component scores for the retained PCA modes) and the predictand will be forced to be monotonic. This flag is not used if configuration selection is 3 but a value must always be given.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: ANN_parallel_flag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Enable simple parallelization across processor cores? If &quot;Y&quot; then foreach loop is used to perform cross-validation for the two neural network prediction systems (monotone artificial neural network, and monotone composite quantile regression neural network)</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    4
                                    <span class="p">(variable name: num_cores)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If ANN_parallel_flag = &quot;Y&quot; (ignored otherwise): set to number of processor cores across which neural network cross-validation tasks will be distributed. There is no default for this value; the user must find the correct value as per the M
                                    <span class="s11">4</span>
                                    installation instructions in step 5 of Section 4.1.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># CONTROL SVM CONFIGURATION:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    2
                                    <span class="p">(variable name: SVM_config_selection)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Set AutoML hyperhyperparameter for support vector machine: if 1, tune gamma, epsilon, and cost; if 2 tune epsilon and cost only; if 3 no automated tuning is performed and hyperparameters are hard-coded by experienced user at corresponding location within the R module implementing SVM. Experimentation with canonical statistical WSF use-cases in the western US suggested that 1 may tend to excessively overtrain; the default configuration is 2, using the default gamma value in the next run control parameter below.</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    0.2112
                                    <span class="p">(variable name: fixedgamma)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">If SVM_config_selection = 2 (ignored otherwise): specify the fixed value of gamma to use in the support vector machine if partial tuning is selected. Default value is 0.2112 (or similar).</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># CONTROL SOME OUTPUTS:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: PC1form_plot_flag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Create scatter plots of observed leading-mode principal component versus observed flow and prediction flow, for each of the six models in M
                                    <span class="s11">4</span>
                                    ? Apart from producing or not producing this plot, the choice has no impact on M
                                    <span class="s11">4</span>
                                    functionality. In general, this is often left to “Y”. However, users should be aware of some complexities in this graphic.
                                </p>
                                <p class="s2" style="padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If more than one PCA mode is retained, this plot is generally not very meaningful. Also, bear in mind that if feature optimization, i.e., the genetic algorithm, is employed, then in general each of the six models in M
                                    <span class="s11">4</span>
                                    will use (1) different input variables and (2) different PCA modes (unless MaxModes is 1), so each model will in this case have its own set of observed PC1 values.
                                </p>
                                <p class="s10" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: PC12form_plot_flag)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">Create contour plots of model predictions as a function of the first and second PCA modes, for each model? This will only be done if the second PCA mode is available, e.g., retained by the genetic algorithm for that model. Note also that if more than two PCA modes are retained, the meaning of this plot may be difficult to interpret.</p>
                                <p style="padding-top: 8pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    In the third and final section of the run control file, the user provides information about how they would like a forecast run to proceed. Correctly formatted values must always be provided; however, M
                                    <span class="s4">4</span>
                                    ignores the values of these parameters and flags if RunTypeFlag = “BUILD” (see above). Setting these values requires some hands-on work from the user, but most of the values are actually determined by (i) what the user asked for during the build phase and/or (ii) what M
                                    <span class="s4">4</span>
                                    produced at the end of that build phase; in general, little to no user judgement is required.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;"># RUN PARAMETERS FOR FORECASTING (Used only if RunTypeFlag = &quot;FORECAST&quot;)</p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;"># SPECIFY WHICH PCA MODES TO RETAIN:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1,2
                                    <span class="p">(variable name: PCSelection_Frwrd_LR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Specify which PCA modes to retain for linear regression (PCR); 1=leading mode, 2=second mode, etc.; if there is more than one mode, separate by commas. It is a combination of up to the first four PCs (as the current M
                                    <span class="s11">4</span>
                                    version functionality is limited by design to a maximum of four PCA modes, as discussed above), determined by what was either manually specified, or determined by the genetic algorithm, during model-building for PCR. Recall that the leading PCA mode is always retained but other modes may be skipped. For example, if only the leading PCA is retained, then the value of PCSelection_Frwrd_LR is 1, but if the first, second, and fourth PCA modes are retained then the value is 1,2,4. Recall also that the maximum number of PCA modes available is also limited by the number of input variables the user provided in the input data file (again see above).
                                </p>
                                <p class="s2" style="padding-top: 5pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If automatic feature optimization wasn’t used during model training (i.e., if GeneticAlgorithmFlag was set to “N” during the build phase) then the value is exactly identical to the value the user specified for ManualPCSelection in the build phase, and may simply be copied and pasted from there.</p>
                                <p class="s2" style="padding-top: 5pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If automatic feature optimization was used during model training (i.e., if GeneticAlgorithmFlag was set to “Y” during the build phase) it’s a little more complicated. In that case, the value of PCSelection_Frwrd_LR is taken by the user from the optimal genetic sequence stored by M
                                    <span class="s11">4</span>
                                    at the end of model training. This is the last line (‘Best Solution’) in the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCR.txt (
                                    <b>not</b>
                                    GA_RunLog_PCR.txt, which is something slightly different); open the example file in the M
                                    <span class="s11">4</span>
                                    distribution and take a look.
                                </p>
                                <p class="s2" style="padding-top: 10pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">That genetic algorithm output information in GA_RunSummary_PCR.txt is used as follows. The first M entries in this binary sequence switch on (1) or off (0) predictor variates, going left to right, as listed in the input data file</p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    MMPEInputData_ModelBuildingMode.txt, which are of course the same M input variables listed in the same order in the forecast-phase data file, MMPEInputData_ForecastingMode.txt (we return to this below). There are then none, one, or two more binary digits in the sequence after those first M bits, depending on how many PCA modes were specified by the user during the build phase for possible retention. The leading PCA mode is always retained, so it is not encoded for here, and if MaxModes was set to 1 by the user during the build phase then the ‘Best Solution’ genetic sequence in GA_RunSummary_PCR.txt is only M binary digits long, and PCSelection_Frwrd_LR is simply set by the user to 1. If MaxModes was set to 2 by the user during the build phase then the ‘Best Solution’ genetic sequence in GA_RunSummary_PCR.txt is M+1 binary digits long, with the (M+1)
                                    <span class="s11">th</span>
                                    digit indicating whether the second PCA mode was retained (1) or not (0). If it’s 1, then the user would enter 1,2 for PCSelection_Frwrd_LR because both the first and second modes are retained. If it’s 0, then the user would enter 1 for PCSelection_Frwrd_LR because only the first mode is retained. The idea is similar for MaxModes = 3 or 4. For example, if the user set MaxModes = 4 in the build phase, then the ‘Best Solution’ genetic sequence in GA_RunSummary_PCR.txt is M+3 binary digits long, with the (M+1)
                                    <span class="s11">th</span>
                                    digit indicating whether the second PCA mode was retained (1) or not (0), the (M+2)
                                    <span class="s11">th</span>
                                    digit indicating whether the third PCA mode was retained (1) or not (0), and the (M+3)
                                    <span class="s11">th</span>
                                    digit indicating whether the fourth PCA mode was retained (1) or not (0). Say if MaxModes = 4, and if the last three digits in the binary string turned out to be 1 0 1, and recalling that the leading PCA mode is always used, then the user would set PCSelection_Frwrd_LR to 1,2,4.
                                </p>
                                <p class="s2" style="padding-top: 5pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Further instructions are given in Section 4.3 below, and examples are provided in the M
                                    <span class="s11">4</span>
                                    distribution files. There is no default value for PCSelection_Frwrd_LR; the user must provide the correct value, and the value shown here is just a randomly chosen example from a past M
                                    <span class="s11">4</span>
                                    run. It may look complicated at first but it’s actually really easy once you get the hang of it.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: PCSelection_Frwrd_QR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above in PCSelection_Frwrd_LR, but for quantile regression (PCQR) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCQR.txt. There is no default value for PCSelection_Frwrd_QR; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1,2
                                    <span class="p">(variable name: PCSelection_Frwrd_mANN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above with PCSelection_Frwrd_LR, but for the monotone artificial neural network (PCANN). In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCANN.txt. There is no default value for PCSelection_Frwrd_mANN; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: PCSelection_Frwrd_MCQRNN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;text-align: left;">As above with PCSelection_Frwrd_LR, but for the monotone composite quantile</p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    regression neural network (PCMCQRNN). In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build- phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCMCQRNN.txt. There is no default value for PCSelection_Frwrd_MCQRNN; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: PCSelection_Frwrd_RF)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above with PCSelection_Frwrd_LR, but for random forests (PCRF). In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCRF.txt. There is no default value for PCSelection_Frwrd_RF; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    1
                                    <span class="p">(variable name: PCSelection_Frwrd_SVM)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above with PCSelection_Frwrd_LR, but for the support vector machine (PCSVM). In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCSVM.txt. There is no default value for PCSelection_Frwrd_SVM; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;"># SPECIFY WHICH INPUT VARIABLES TO RETAIN: 1 TO KEEP, 0 TO OMIT, POSITIONS CORRESPOND TO LEFT-TO-RIGHT CONSECUTIVE LOCATIONS IN INPUT DATASET</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 79pt;text-indent: 0pt;text-align: left;">
                                    &quot;0 1 0 1 0 1&quot;
                                    <span class="p">(variable name: VariableSelection_Frwrd_LR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Specify which of the input variables to use for linear regression, i.e., PCR. Use quotation marks, and single spaces with no commas between each binary digit in the string, as shown in the example above. This length-M binary string corresponds to the M input variables listed left-to-right in the build-phase data file, MMPEInpuData_ModelBuildingMode.txt, and which are of course the same M input variables listed in the same order in the forecast-phase data file, MMPEInputData_ForecastingMode.txt. 1 indicates retain the variable, and 0 indicates that the variable will not be retained.
                                    <span class="s13">.</span>
                                </p>
                                <p style="text-indent: 0pt;text-align: left;">
                                    <br>
                                </p>
                                <p class="s2" style="padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">If automated feature optimization wasn’t used, i.e., if GeneticAlgorithmFlag was set to “N” during the build phase, all the variables in the input data file are retained (see above), and the user simply sets VariableSelection_Frwrd_LR to a length-M string of 1s. For instance, if MMPEInpuData_ModelBuildingMode.txt contained seven variables, then set VariableSelection_Frwrd_LR to “1 1 1 1 1 1 1”.</p>
                                <p class="s2" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If automated feature optimization was used, i.e., if GeneticAlgorithmFlag was set to “Y” during the build phase, then VariableSelection_Frwrd_LR is set to the first M digits in the binary string at the bottom of the file, GA_RunSummary_PCR.txt, that was saved by M
                                    <span class="s11">4</span>
                                    during the build phase. (This string may or may not be of length &gt; M, depending on whether MaxModes was set to &gt; 1 or not; for setting VariableSelection_Frwrd_LR, ignore everything after the first M digits.) That is, the first M entries in the binary string at the end of GA_RunSummary_PCR.txt can simply be copied-and-pasted into VariableSelection_Frwrd_LR .
                                </p>
                                <p class="s2" style="padding-top: 5pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    Further instructions are given in Section 4.3 below, and examples are provided in the M
                                    <span class="s11">4</span>
                                </p>
                                <p class="s2" style="padding-top: 2pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    distribution files. See also description of variable PCSelection_Frwrd_LR above. There is no default value for VariableSelection _Frwrd_LR; the user must provide the correct value. The value shown here is just a randomly chosen example from a past run of M
                                    <span class="s11">4</span>
                                    .
                                </p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: justify;">
                                    &quot;0 1 0 1 1 0&quot;
                                    <span class="p">(varianble name: VariableSelection_Frwrd_QR)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: justify;">
                                    As above in VariableSelection_Frwrd_LR, but for quantile regression (PCQR) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCQR.txt.
                                </p>
                                <p class="s2" style="padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: justify;">There is no default value for VariableSelection_Frwrd_QR); the user must provide the correct value.</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;0 0 1 1 1 1&quot;
                                    <span class="p">(variable name: VariableSelection_Frwrd_mANN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above in VariableSelection_Frwrd_LR, but for the monotone artificial neural network (PCANN) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCANN.txt. There is no default value for VariableSelection_Frwrd_mANN); the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;0 1 0 1 0 1&quot;
                                    <span class="p">(variable name: VariableSelection_Frwrd_MCQRNN)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above in VariableSelection_Frwrd_LR, but for the monotone composite quantile regression neural network (PCMCQRNN) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build- phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCMCQRNN.txt. There is no default value for VariableSelection_Frwrd_MCQRNN; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;0 0 0 1 1 0&quot;
                                    <span class="p">(variable name: VariableSelection_Frwrd_RF)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above in VariableSelection_Frwrd_LR, but for the random forests model (PCRF) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCRF.txt. There is no default value for VariableSelection_Frwrd_RF; the user must provide the correct value
                                    <span class="p">.</span>
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;0 0 0 1 0 1&quot;
                                    <span class="p">(variable name: VariableSelection_Frwrd_SVM)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As above in VariableSelection_Frwrd_LR, but for the support vector machine (PCSVM) instead. In this case, if the genetic algorithm was employed in the build phase, the optimal genetic sequence is provided by the build-phase M
                                    <span class="s11">4</span>
                                    output file, GA_RunSummary_PCSVM.txt. There is no default value for VariableSelection_Frwrd_SVM; the user must provide the correct value.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: justify;"># HOUSEKEEPING FOR MONOTONIC NEURAL NETS:</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: ANN_monotone_flag_Frwrd)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;">Was the monotonicity constraint enabled for the neural networks (PCANN, PCMCQRNN) in the build run? Enter “Y” is if it was, and “N” otherwise; use caps and keep the quotation marks. If default values were used during the build phase, this will always be “Y”.</p>
                                <p class="s10" style="padding-top: 4pt;padding-left: 80pt;text-indent: 0pt;line-height: 108%;text-align: left;"># SPECIFY WHETHER ENSEMBLE MEAN FORECAST DISTRIBUTION IS TO BE GENERATED AND IF SO FROM WHICH MODELS</p>
                                <p class="s10" style="padding-top: 8pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;Y&quot;
                                    <span class="p">(variable name: Ensemble_flag_frwrd)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Create an multi-model ensemble mean forecast distribution by calculating the average of the forecast distributions (specifically, the 10, 30, 70, and 90% exceedance volumes and the best-estimate prediction) from the individual models in the forward runs? If set to &quot;N&quot; then only the forecast distributions from the individual models are written to file; if &quot;Y&quot; then the ensemble mean is additionally calculated and written to file. Although this flag is functional and the user can set it to “N” if they wish, it is a largely an artefact from an early development version of M
                                    <span class="s11">4</span>
                                    and in general it should always be set to “Y”. Use caps and don’t drop the quotation marks.
                                </p>
                                <p class="s10" style="padding-top: 7pt;padding-left: 80pt;text-indent: 0pt;text-align: left;">
                                    &quot;AUTO&quot;
                                    <span class="p">(variable name: Ensemble_type_frwrd)</span>
                                </p>
                                <p class="s2" style="padding-top: 1pt;padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    If Ensemble_flag_frwrd = &quot;Y&quot; (ignored otherwise): &quot;ALL&quot; creates an ensemble mean from the full default model set (PCR-BC, PCQR, PCANN-BC, PCMCQRNN, PCRF-BC, PCSVM-BC), &quot;AUTO&quot; does the same but removes any individual ensemble members that may have been pruned from the final build-phase ensemble if M
                                    <span class="s11">4</span>
                                    ’s negativity check algorithm was switched on by the user during the build phase by setting AutoEnsembleFlag to “Y”.
                                </p>
                                <p class="s2" style="padding-left: 79pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Whether such exclusions were needed, and if so which models were excluded, was recorded by M
                                    <span class="s11">4</span>
                                    in the build-phase output file, AutomatedEnsembleMemberExclusions.csv, which M
                                    <span class="s11">4</span>
                                    reads in during forecast runs if Ensemble_type_frwrd is set to “AUTO”. Generally speaking, the default for AutoEnsembleFlag is “Y” (see above) and the default for Ensemble_type_frwrd is “AUTO”. Use caps and don’t drop the quotation marks.
                                </p>
                            </li>
                            <li data-list-text="4.2.3">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Output files</p>
                                <ol id="l16">
                                    <li data-list-text="4.2.3.1">
                                        <p class="s14" style="padding-top: 9pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Output file naming conventions</p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">Some file name conventions used here are:</p>
                                        <ul id="l17">
                                            <li data-list-text="">
                                                <p style="padding-top: 9pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    PCR_BC: refers to output from the linear regression model within M
                                                    <span class="s4">4</span>
                                                    (i.e., principal component regression, given the PCA preprocessing used in all six modeling systems within M
                                                    <span class="s4">4</span>
                                                    ) when Box- Cox transform-space residuals are used as the basis for prediction interval estimation capable of handling non-Gaussian and heteroscedastic error distributions.
                                                </p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    PCR: refers to output from the linear regression model in M
                                                    <span class="s4">4</span>
                                                    for (i) outputs that are specific to the use of classical Gaussian stationary prediction intervals as in most traditional applications of PCR to western US WSF, in contrast to the Box-Cox (BC) space method described above, and (ii) outputs which are the same irrespective of the method used for prediction interval estimation. The deterministic regression model and associated properties, like which input variables and PCA modes are retained for instance, are identical between the two cases, so for the most part that kind of information is not repeated separately for PCR_BC. Similarly, the best-estimate models and best-estimate predictions are identical between PCR and PCR_BC. See also detailed descriptions of “CONTROL ENSEMBLE GENERATION” flags in MMPE_RunControlFile.txt portion of Section 4.2.2 above.
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-top: 3pt;padding-left: 44pt;text-indent: -18pt;text-align: left;">PCRF_BC: as for PCR_BC above, but for random forests.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">PCRF: as for PCR above, but for random forests.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">PCSVM_BC: as for PCR_BC above, but for the support vector machine.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">PCSVM: as for PCR above, but for the support vector machine.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">PCANN_BC: as for PCR_BC above, but for the monotone artificial neural network.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">PCANN: as for PCR above, but for the monotone artificial neural network.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">PCQR: refers to output from linear quantile regression. Because quantile regression inherently generates non-Gaussian and heteroscedastic prediction intervals, there is no separate PCQR_BC.</p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">PCMCQRNN: refers to output from the monotone composite quantile regression neural network. Because quantile regression techniques inherently generate non-Gaussian and heteroscedastic prediction intervals, there is no separate PCMCQRNN_BC.</p>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.2.3.2">
                                        <p class="s14" style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Output files created by build-step M
                                            <span class="s15">4</span>
                                            runs irrespective of whether feature optimization is used
                                        </p>
                                        <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            The following are output files created by M
                                            <span class="s4">4</span>
                                            during model-building runs (that is, if RunTypeFlag = “BUILD”), regardless of whether the genetic algorithm is switched on or off by the user in the run control file, that is, for GeneticAlgorithmFlag = “Y” or “N” alike.
                                        </p>
                                        <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Files and their descriptions are grouped below by content. Many of these files contain information about the build-phase results that are required by M
                                            <span class="s4">4</span>
                                            during the forecasting-mode runs but which may not otherwise be of much interest to the user, whereas other files contain key results.
                                        </p>
                                        <ul id="l18">
                                            <li data-list-text="">
                                                <p style="padding-top: 7pt;padding-left: 44pt;text-indent: -18pt;text-align: left;">Standard performance metrics:</p>
                                                <ul id="l19">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;text-align: left;">These files give a standard set of fit measures for each of the models, following the naming convention described in Section 4.2.3.1 above.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l20">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: left;">ensemble_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_BC_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_BC_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_BC_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_BC_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_StandardPerformanceReportingSuite.csv</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 3pt;padding-left: 79pt;text-indent: -18pt;text-align: left;">The core fit measures provided in these files for each model are:</p>
                                                        <ul id="l21">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    Cross-validated (out-of-sample) coefficient of determination (R
                                                                    <span class="s4">2</span>
                                                                    ), a standard deterministic model fit metric, where 1 is perfect and lower values denote progressively poorer performance with 0 indicating no ability to reproduce the data
                                                                </p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: justify;">Cross-validated (out-of-sample) root mean square error (RMSE), a standard deterministic model fit metric, where 0 is perfect and higher values denote progressively poorer performance</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">Cross-validated (out-of-sample) ranked probability skill score (RPSS), a standard probabilistic model fit metric which in some sense gives an estimate of how good the prediction intervals are, where 1 is perfect and lower values denote progressively poorer performance; the value is based on post facto probabilities estimated on the assumption of either a homoscedastic Gaussian error distribution centered around the best estimate with a standard deviation equal to the RMSE (e.g., for PCR) or similarly but in Box-Cox transform space (e.g., for PCR_BC) which allows for heteroscedastic and non-Gaussian residuals; the exception is the two modeling systems based on quantile regression-type calculations, where the probability estimates are intrinsic to the technique</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: left;">Cross-validated (out-of-sample) categorical hit rates for the lowermost (HR-L), middle (HR-M), and uppermost (HR-H) terciles of observations, where perfect values are 1, and any value &gt; 1/3 indicates the presence of skill relative the climatology forecast (to borrow a term from the numerical weather prediction community)</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 80pt;text-indent: -18pt;line-height: 14pt;text-align: left;">Some subsidiary in-sample fit measures also included in these files are:</p>
                                                        <ul id="l22">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    Whether the best-estimate prediction is negative-valued (i.e., for runoff volume, nonphysical) for any sample during the training period (BE&lt;0?); note that for the ensemble mean, and if the option to use the nonnegativity pruning algorithm is selected by the user as it is in the default M
                                                                    <span class="s4">4</span>
                                                                    application, this metric is evaluated after any such ensemble member removal has taken place
                                                                </p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    Whether the lowermost prediction bound considered (90% exceedance probability flow volume) is negative-valued (i.e., for runoff volume, nonphysical) for any sample during the training period (PB&lt;0?): note that for the ensemble mean, and if the option to use the negativity pruning algorithm is selected by the user as it is in the default M
                                                                    <span class="s4">4</span>
                                                                    application, this metric is evaluated after any possible ensemble member removal has taken place
                                                                </p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    The forecast distribution, expressed as categorical bin counts where the bands are segmented according to the prediction intervals considered (between the minimum and the 90% exceedance flow, FD 100-90; between the 90% and 70% exceedance flows, FD 90-70; between the 70% exceedance flow and the best estimate, FD 70-BE; between the best estimate and the 30% exceedance flow, FD BE-30; between the 30% and 10% exceedance flows, FD 30-10; and between the 10% exceedance flow the maximum, FD 10-0); where the expected frequencies are the number of samples,
                                                                    <i>N</i>
                                                                    , multiplied by 0.1, 0.2, 0.2, 0.2, 0.2, and 0.1 respectively; these statistics do not seem particularly stable for small
                                                                </p>
                                                                <p style="padding-top: 2pt;padding-left: 116pt;text-indent: 0pt;line-height: 108%;text-align: justify;">sample sizes often encountered in canonical WSF use-cases and might not be accorded a great deal of meaning but are provided as additional information nevertheless</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: left;">The maximum absolute error and the maximum relative (percentage) error and the years (or sample indices) in which they occur; 0 is a perfect model; the years during which the maximum absolute and relative errors occur are not necessarily the same.</p>
                                                                <p style="text-indent: 0pt;text-align: left;">
                                                                    <br>
                                                                </p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;text-align: left;">Augmented performance metrics:</p>
                                                <ul id="l23">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">For each of the models, and following the naming convention described in Section 4.2.3.1, these files give both the in-sample and cross-validated (out-of-sample) values for a key subset of the metrics described above.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l24">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: left;">enemsble_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_BC_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_BC_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_BC_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_BC_AdditionalReportingSuite.csv</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 80pt;text-indent: -18pt;text-align: left;">
                                                            The measures considered are R
                                                            <span class="s4">2</span>
                                                            , RMSE, and the tercile hit rates (HR-L, HR-M, HR-H) (see above for descriptions).
                                                        </p>
                                                        <p style="text-indent: 0pt;text-align: left;">
                                                            <br>
                                                        </p>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;text-align: left;">Model predictions:</p>
                                                <ul id="l25">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 80pt;text-indent: -18pt;line-height: 107%;text-align: left;">For each of the models, and following the naming convention described in Section 4.2.3.1, these files give the forecast distributions for each sample time over the training period for each of the constituent modeling systems. Following typical practice, in- sample values are provided, which speak primarily to the ability of the model to reproduce historical data and can be helpful for diagnosing and understanding model dynamics. For measures of predictive performance, focus instead on cross-validated (out-of-sample) performance statistics mentioned above. Out-of-sample predictions can also be found in some of the files listed under “Additional Information” below.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 80pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l26">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;text-align: left;">ensemble_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 116pt;text-indent: -18pt;text-align: left;">PCANN_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_predictions_BC.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_predictions_BC.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 3pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_predictions_BC.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_predictions.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_predictions_BC.csv</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            The first column is a sequential index, the second column is the user-specific sample index (assumed to be the year), the third column is the observed predictand (flow volume), the fourth column gives the best estimate of the predictand, and the next four columns give the prediction intervals as the 90%, 70%, 30%, and 10% exceedance flow volumes (i.e., 10
                                                            <span class="s4">th</span>
                                                            , 30
                                                            <span class="s4">th</span>
                                                            , 70
                                                            <span class="s4">th</span>
                                                            , and 90
                                                            <span class="s4">th</span>
                                                            percentiles of the error distribution, respectively).
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">An additional file compiles the best estimates for each of the models into a single table:</p>
                                                        <ul id="l27">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: left;">BestEstimatesAllModels.csv</p>
                                                                <p style="text-indent: 0pt;text-align: left;">
                                                                    <br>
                                                                </p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;text-align: left;">Principal component analysis (PCA) results:</p>
                                                <ul id="l28">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;text-align: left;">For each of the models, and following the naming convention described in Section 4.2.3.1, these files give outputs from the PCA data preprocessing step.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">If the user specified that genetic algorithm-based feature optimization was to be used, the PCA results will in general be different for each of the six constituent modeling systems, as each is independently optimized and will typically retain different input variables from the candidate pool, leading to different PCA results.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">If feature optimization was not used, then the same data were used for all the models, and the PCA results will be identical between the constituent modeling systems, and many of the files below will therefore have identical content.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l29">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_eigenvector.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_MeanOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCANN_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_MeanOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNN_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_eigenvector.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_MeanOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_eigenvector.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_MeanOfEachRetainedInputVariable.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCR_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_eigenvector.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 3pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_MeanOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRF_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_eigenspectrum.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_eigenvector.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_MeanOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVM_PCtimeseries.csv</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: justify;">PCSVM_StdevOfEachRetainedInputVariate.csv</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: justify;">As the foregoing list shows, there is a set of five files for the final retained predictor set (if the genetic algorithm is used) or for the full set of predictors in the input data file (if the genetic algorithm is not used) for each model:</p>
                                                        <ul id="l30">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">One file (modelabbreviation_eigenspectrum.csv) gives the eigenvalues for each mode, expressed in terms of the percentage of dataset variance explained by each PCA mode</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">One file (modelabbreviation_eigenvector.csv) gives the eigenvectors (loading patterns) expressed as a matrix: each column contains an eigenvector, with the leading mode in the left-most column, and each row corresponds to a different variable, sorted in the same order as in the input data file (but omitting variables not retained if the genetic algorithm was invoked).</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">One file (modelabbreviation_PCtimeseries.csv) give the principal component time series (scores) expressed as a matrix, where the rows correspond to PCA modes (1 is leading mode, etc.) and the columns correspond to sequential sampling times</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">One file (modelabbreviation_MeanOfEachRetainedInputVariate.csv) captures the mean of each of the retained input variables, as calculated during the process of standardizing each of those variables to zero mean and unit variance, such that PCA is performed on the correlation rather than covariance matrix.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">One file (modelabbreviation_StdevOfEachRetainedInputVariate.csv) captures the standard deviation of each of the retained input variables, as calculated during the process of standardizing each of those variables to zero mean and unit variance, such that PCA is performed on the correlation rather than covariance matrix.</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 106%;text-align: left;">
                                                            The primary purpose of these PCA output files is to store information that will be used later in a forecasting run of M
                                                            <span class="s4">4</span>
                                                            .
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            They also can provide valuable diagnostic information for intermediate- to advanced- level users of M
                                                            <span class="s4">4</span>
                                                            and, in particular, support explainable machine learning in the context of WSF (see Appendices).
                                                        </p>
                                                        <p style="text-indent: 0pt;text-align: left;">
                                                            <br>
                                                        </p>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;text-align: left;">Models:</p>
                                                <ul id="l31">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            For each of the models, and following the naming convention described in Section 4.2.3.1, these files define the final supervised learning models for each of the modeling systems within the current version of M
                                                            <span class="s4">4</span>
                                                            .
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l32">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: justify;">PCANNmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 3pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCMCQRNNmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRFmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCRmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCSVMmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQRmodel.Rdata</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">PCQR_model_summary.txt</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">As noted in Section 4.2.1, the quantile regression module provides slightly different output format than the other five constituent modeling systems within the current version. See also comments in the R file containing the quantile regression function.</p>
                                                        <p style="text-indent: 0pt;text-align: left;">
                                                            <br>
                                                        </p>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;text-align: left;">Additional information:</p>
                                                <ul id="l33">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;text-align: left;">Additional files providing information of potential interest to users are:</p>
                                                        <ul id="l34">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    AutomatedEnsembleMemberExclusions.csv shows which, if any, of the constituent modeling systems were excluded from the final ensemble by the pruning algorithm, if used (see run control flags around automatic ensemble generation above), for contributing to a negative-valued ensemble mean prediction interval at any sample time during the training period (see above and appendices for more details); for default applications, this information is also used by M
                                                                    <span class="s4">4</span>
                                                                    during a subsequent forecast run.
                                                                </p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">BoxCoxPredictionIntervalsSwitches.csv shows which, if any, of PCR_BC, PCSVM_BC, PCRF_BC, and/or PCANN_BC have had their prediction intervals replaced by those from PCR, PCSVM, PCRF, and/or PCANN, respectively. This occurs in rare cases, on a model-by-model basis, when the Inf check algorithm has been tripped has been tripped for that model; see the description of AutoEnsembleFlag in Section 4.2.2 above. Given that the best estimates are always the same between PCR and PCR_BC, PCRF and PCRF_BC, PCSVM and PCSVM_BC, and PCANN and PCANN_BC, in such cases PCR_BC, PCSVM_BC, PCRF_BC, and/or PCANN_BC are made fully identical to PCR, PCSVM, PCRF, and/or PCANN. Of course these changes, on the rare occasion they occur, are also reflected in the calculation of the multi-model ensemble mean forecast distribution.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">residuals_all_models.csv gives the training-period in-sample and cross-validated (LOOCV) residuals for the best-estimate predictions from each constituent modeling system and from the ensemble mean.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                                    error_message_log.txt logs output messages, if the option to write it is turned on by the user in the run control file (as in the default case; see above); this should in general be briefly consulted after a M
                                                                    <span class="s4">4</span>
                                                                    run is complete; most warnings and other information can usually be ignored, but in the event of a crash the errors logged here can be useful for diagnosis, particularly for experienced R and RStudio users.
                                                                </p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">The files PCANN_AdditionalDataForPostProcessing.csv, PCANN_BC_AdditionalDataForPostProcessing.csv, PCMCQRNN_AdditionalDataForPostProcessing.csv, PCQR_AdditionalDataForPostProcessing.csv,</p>
                                                                <p style="padding-top: 2pt;padding-left: 116pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    PCR_AdditionalDataForPostProcessing.csv, PCR_BC_AdditionalDataForPostProcessing.csv, PCRF_AdditionalDataForPostProcessing.csv, PCRF_BC_AdditionalDataForPostProcessing.csv, PCSVM_AdditionalDataForPostProcessing.csv, and PCSVM_BC_AdditionalDataForPostProcessing.csv each provide information that is used by the Persephone UI (not included in this M
                                                                    <span class="s4">4</span>
                                                                    distribution) and may not be of much interest to most users, but also includes both the in-sample and out- of-sample best-estimate predictions and residuals for each of the constituent modeling systems.
                                                                </p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">
                                                            Additional files providing information that in general may be of little interest to users but is required in subsequent M
                                                            <span class="s4">4</span>
                                                            forward runs using the built models:
                                                        </p>
                                                        <ul id="l35">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">AutomaticallyOptimizableHyperparameters.csv records some hyperparameter tuning results.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">BoxCoxInfo.csv records some automated Box-Cox transform parameter estimation results.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">MontonicityConstraintRequiredAdjustments.csv records some AutoML outcomes for the neural networks.</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;line-height: 108%;text-align: left;">QcritandN.csv records information about the terciles of the observed predictand over the training period, used for some statistical performance diagnostics, and the number of samples used in the build phase.</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.2.3.3">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Additional output files created by build-step M
                                            <span class="s15">4</span>
                                            runs if automated feature optimization is used
                                        </p>
                                        <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            The following are additional output files created by M
                                            <span class="s4">4</span>
                                            during model-building runs (that is, if RunTypeFlag = “BUILD”) only if the genetic algorithm is switched on by the user in the run control file, that is, for GeneticAlgorithmFlag = “Y”.
                                        </p>
                                        <ul id="l36">
                                            <li data-list-text="">
                                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;text-align: left;">Genetic algorithm status reports piped to run logs:</p>
                                                <ul id="l37">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">For each of the models, and following the naming convention described in Section 4.2.3.1, these files give intermediate outputs from the genetic algorithm as it evolves toward an optimal solution.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 108%;text-align: left;">This information is not used in subsequent forecast operations and will usually be of limited interest to the user, but there may be cases where a user may find it instructive to review the run logs to determine if a set of features (combination of retained predictor variates, and principal component time series derived from that retained set of candidates) other than the final optimized feature set might provide similar performance but offer other more desirable characteristics (e.g., retaining more predictors, giving better operational redundancy in operations) in accordance with the hydrologist’s professional judgement.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 80pt;text-indent: -18pt;line-height: 13pt;text-align: left;">The files are:</p>
                                                        <ul id="l38">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCANN.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 116pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCMCQRNN.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 116pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCQR.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 3pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCR.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCRF.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunLog_PCSVM.txt.</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-top: 1pt;padding-left: 43pt;text-indent: -18pt;text-align: left;">Final genetic algorithm results:</p>
                                                <ul id="l39">
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 1pt;padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">For each of the models, and following the naming convention described in Section 4.2.3.1, these files give the final results from the genetic algorithm and describe which input variables and (if MaxModes is set by the user to &gt; 1 in the run control file, PCA modes) are retained in each of the six final modeling systems.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            The file is small. It records some of the genetic algorithm parameters set by the user in the run control file, or which are hard-wired in the M
                                                            <span class="s4">4</span>
                                                            code but accessible to expert users. The most important portion is the final line, which provides key information for the user when setting up the run control file for subsequent forecasts using the final build-phase models. This information is encoded in the string of binary digits (that is, the string of 0s and 1s) following “Best Solution : “
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                            The first
                                                            <i>M</i>
                                                            binary digits in this string indicate which of the input variables in the candidate pool supplied by the user to M
                                                            <span class="s4">4</span>
                                                            in MMPEInputData_ModelBuildingMode.txt are retained in the final model. For a detailed explanation, see the description given above in Section 4.2.2 for how to set VariableSelection_Frwrd_LR in MMPE_RunControlFile.txt . To recap, reading this string from left to right, 1 means the input predictor variable is kept and 0 means the variable isn’t retained, where the order is identical to the input predictor variables as listed left to right in MMPEInputData_ModelBuildingMode.txt. Think of the string as being a template placed over the headers in MMPEInputData_ModelBuildingMode.txt: 1 means keep that variable, 0 means it’s discarded.
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 13pt;text-align: left;">
                                                            If MaxModes is set by the user to 1 in the run control file, the binary string is only
                                                            <i>M</i>
                                                        </p>
                                                        <p style="padding-left: 79pt;text-indent: 0pt;text-align: left;">digits long.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            If MaxModes is set by the user to 2
                                                            <span class="s16"></span>
                                                            <span class="s17"></span>
                                                            <i>Q</i>
                                                            <span class="s16"></span>
                                                            <span class="s17"></span>
                                                            4 in the run control file, there will be
                                                            <i>Q</i>
                                                            -1 additional digits in the binary string after the first
                                                            <i>M</i>
                                                            digits which encode which of the PCA modes (beyond the first mode, which is always retained and therefore not explicitly encoded for here) is retained. For a detailed explanation, see the description given above in Section 4.2.2 for how to set PCSelection_Frwrd_LR in MMPE_RunControlFile.txt . To summarize, the (
                                                            <i>M</i>
                                                            +1)th through
                                                            <i>M</i>
                                                            +
                                                            <i>Q</i>
                                                            -1)th digits in the binary string, respectively, encode whether the
                                                            <i>Q</i>
                                                            th mode is retained.
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;text-align: left;">The files are:</p>
                                                        <ul id="l40">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCANN.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCMCQRNN.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCQR.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCR.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCRF.txt</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-top: 1pt;padding-left: 115pt;text-indent: -18pt;text-align: left;">GA_RunSummary_PCSVM.txt.</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.2.3.4">
                                        <p class="s14" style="padding-top: 9pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Unsaved graphical outputs</p>
                                        <ul id="l41">
                                            <li data-list-text="">
                                                <p style="padding-top: 9pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    M
                                                    <span class="s4">4</span>
                                                    generates a large number of plots during the build phase. They are intended primarily for diagnostic purposes and are not written to file. While users are encouraged to explore these
                                                </p>
                                                <p style="padding-top: 2pt;padding-left: 44pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                    graphics, some of their outputs are complex or come with caveats, and the output information that most users will want is contained in the .csv and .txt files written by M
                                                    <span class="s4">4</span>
                                                    .
                                                </p>
                                                <ul id="l42">
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            A partial exception may be a multi-panel diagnostics plot; these are produced for each of the constituent modeling systems but the diagnostics plots for the final multi-model ensemble mean prediction will generally be of the most interest to users. An example is given below for an application of M
                                                            <span class="s4">4</span>
                                                            to the Gila River.
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                            The top panel gives historical in-sample performance of M
                                                            <span class="s4">4</span>
                                                            during training. The red dots connected by a thick red line are the observations, the thick black line is the M
                                                            <span class="s4">4</span>
                                                            ensemble best-estimate prediction, and the dashed gray lines are the prediction intervals (dashed gray lines give estimated 30% and 70% exceedance flows, and solid gray lines give estimated 10% and 90% exceedance flows). A thin red horizontal line indicates zero flow for reference. This multi-model ensemble mean forecast distribution was generated using the default model combination (linear quantile regression and a monotone composite quantile regression neural network with prediction intervals intrinsic to those quantile regression methods; and linear regression, a monotone artificial neural network, random forests, and a support vector machine, each individually using prediction intervals generated in Box-Cox transform space on the basis of the out-of-sample root mean square errors of each model). The nonnegativity check algorithm did not require removal of any ensemble members. Note that the prediction intervals around the ensemble mean prediction, which are the means of the prediction intervals from the six modeling systems, are asymmetric around the best estimate and vary in width from year to year with wider intervals during high-flow years, suggesting heteroscedastic and non-Gaussian error distributions; this is justified by the results in the next two rows of the diagnostics plot (see immediately below).
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">The second row down gives scatter plots of observed vs. predicted flows and of model error as a function of the predicted flow. The diagonal line in the lefthand plot gives the 1:1 ratio that a perfect prediction would follow. These two plots also provide qualitative indications around whether heteroscedasticity is present. In this example, the error variance is obviously larger when flows are high; this nonstationarity in the variance is not unusual in many environmental systems.</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 107%;text-align: left;">
                                                            The third row down gives information around the distributional form of the residuals, including a quantile-quantile plot (a normal distribution would follow the diagonal line), a Shapiro-Wilk test on the normality of the error distribution (
                                                            <i>p</i>
                                                            -value &lt; 0.05 to 0.10 or so is typically taken to indicate that the null hypothesis of a normal distribution can be rejected), and a simple histogram. The error distribution is, in this example, clearly non- Gaussian.
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 79pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                            The fourth row down gives information about linear autocorrelation in the residuals at lags up to 10 (in the canonical WSF use-case this corresponds to a lag of 10 years given the use of annual time series; see example dataset given in Section 4.2.2 above and the accompanying M
                                                            <span class="s4">4</span>
                                                            application examples), where the blue lines indicate rejection of the null hypothesis of no serial dependence at
                                                            <i>p</i>
                                                            = 0.05. Both the autocorrelation and the partial autocorrelation (which can be helpful for localizing autocorrelation to a particular lag) functions are provided. There is little evidence in this particular example for any substantial autocorrelation in model error.
                                                        </p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-top: 3pt;padding-left: 79pt;text-indent: -18pt;line-height: 108%;text-align: left;">The bottom panels give some information on categorical and distributional forecast performance. On the left, the ability of the model to predict above-normal, normal, or below-normal conditions, where the category cutoffs are defined as terciles of the observed predictand, is depicted. A perfect hit rate is 1 (dashed blue line) and the simple climatology forecast is 1/3 (dashed red line), so a result above the red line indicates the presence of prediction skill. The right-most panel breaks up the forecast distribution according to the prediction intervals. Expected performance is shown by the black bars, whereas actual performance is shown by the gray bars; note, however, that these statistics do not seem to be particularly stable for the small sample sizes often encountered in canonical WSF use-cases and might not be accorded a great deal of meaning. The information in the bottom panels is also provided in the set of standard performance measures written to a .csv file as described in Section 4.2.3.2.</p>
                                                        <p class="s14" style="padding-top: 9pt;padding-left: 42pt;text-indent: -36pt;text-align: center;">
                                                            <img src="https://github.com/nrcs-nwcc/nrcs-nwcc.github.io/blob/d7f81a465d855a544629726eb2a370b82b658e5e/assets/img/GilaStandardApplicationDiagnostics.jpeg">
                                                        </p>
                                                        <p style="text-indent: 0pt;text-align: left;">
                                                            <br>
                                                        </p>
                                                    </li>
                                                </ul>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.2.3.5">
                                        <p class="s14" style="padding-top: 3pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                                            Output files created by forecast-step M
                                            <span class="s15">4</span>
                                            runs
                                        </p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">
                                            A forecasting run of M
                                            <span class="s4">4</span>
                                            creates only three files:
                                        </p>
                                        <ul id="l43">
                                            <li data-list-text="">
                                                <p style="padding-top: 9pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">ForwardRunMultiModelEnsembleMeanPrediction.csv gives the best-estimate predictions and associated prediction intervals, generated using the finalized models created during the</p>
                                                <p style="padding-top: 3pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                    preceding build-phase run of M
                                                    <span class="s4">4</span>
                                                    and the new sample fed to M
                                                    <span class="s4">4</span>
                                                    by the user in MMPEInputData_Forecasting.txt, in accordance with the options specified by the user in MMPE_RunControlFile.txt. Recall that the ensemble mean will in general use only a subset of the available models; if AutoEnsembleFlag is “Y” then these will be the six default models, potentially pruned by the automated negativity check algorithm to help ensure nonnegativity of the runoff prediction, as described above. Note also that if the Inf check algorithm was triggered in training, then the prediction intervals in forecasts generated using PCR_BC, PCRF_BC, PCSVM_BC, and/or PCANN_BC are replaced by those in PCR, PCRF, PCSVM, and/or PCANN, respectively (refer back to description of BoxCoxPredictionIntervalsSwitches.csv in Section 4.2.3.2 and description of AutoEnsembleFlag in Section 4.2.2), and this information is propagated into the multi-model ensemble mean prediction.
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    ForwardRunAllModelsPredictions.csv gives the best-estimate predictions and associated prediction intervals for each of the 10 available constituent modeling systems (6 models, plus both conventional and Box-Cox transform space-based prediction intervals for random forests, support vector machine, linear regression, and the monotone artificial neural network, as described above) generated using the finalized models created during the preceding build-phase run of M
                                                    <span class="s4">4</span>
                                                    and the new sample fed to M
                                                    <span class="s4">4</span>
                                                    by the user in MMPEInputData_Forecasting.txt, in accordance with the options specified by the user in MMPE_RunControlFile.txt. Note that if the Inf check algorithm was triggered in training, then the prediction intervals in forecasts generated using PCR_BC, PCRF_BC, PCSVM_BC, and/or PCANN_BC are replaced by those in PCR, PCRF, PCSVM, and/or PCANN, respectively (refer back to description of BoxCoxPredictionIntervalsSwitches.csv in Section 4.2.3.2 and description of AutoEnsembleFlag in Section 4.2.2).
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">error_messages_log.txt as in the build phase as described above; this overwrites the file of the same name created during the build phase.</p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 44pt;text-indent: -18pt;line-height: 14pt;text-align: left;">
                                                    No graphics are produced by M
                                                    <span class="s4">4</span>
                                                    in forecast-step runs.
                                                </p>
                                                <p style="text-indent: 0pt;text-align: left;">
                                                    <br>
                                                </p>
                                            </li>
                                        </ul>
                                    </li>
                                </ol>
                            </li>
                        </ol>
                    </li>
                    <li data-list-text="4.3">
                        <h4 style="padding-left: 44pt;text-indent: -36pt;text-align: left;">Use instructions</h4>
                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            M
                            <span class="s4">4</span>
                            use instructions are provided below and divided into three groups by user type: basic (Section 4.3.1), intermediate (Section 4.3.2), and expert (Section 4.3.3). These instructions necessarily assume that the user has read and understood the preceding sections of this manual and the background provided in the Appendices. Detailed examples distributed in the repo are described below in Section 5.
                        </p>
                        <ol id="l44">
                            <li data-list-text="4.3.1">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Basic instructions</p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The basic instructions provided in this section correspond to a simple, default approach for applying M
                                    <span class="s4">4</span>
                                    to the NRCS WSF canonical use-case (see above). It is intended for users whose primary area of interest and expertise is in hydrology and, in particular, operational water supply forecasting in the American West, but who may have limited knowledge of machine learning or the other technical methods deployed in M
                                    <span class="s4">4</span>
                                    .
                                </p>
                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    This mode of application relies heavily on the AutoML capabilities of M
                                    <span class="s4">4</span>
                                    . Its capabilities as a WSF approach should not be underestimated; based on retrospective and live operational testing conducted so far, this standardized approach appears to consistently out-perform current WSF methods in
                                </p>
                                <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    common use at governmental service-delivery organizations in the US West, and it may be anticipated to yield reasonable results for the WSF canonical use case while maintaining reasonable run times and geophysical explainability of results. That said, we emphasize again (see Disclaimer above) that M
                                    <span class="s4">4</span>
                                    remains a prototype. Additionally, the basic instructions may not provide the best performance available potentially available from M
                                    <span class="s4">4</span>
                                    in a given application, and ideally, it should be viewed as a starting point for learning more about M
                                    <span class="s4">4</span>
                                    and the details of the techniques it deploys, and how to use those methods to maximum advantage.
                                </p>
                                <ol id="l45">
                                    <li data-list-text="4.3.1.1">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                                            Building a M
                                            <span class="s15">4</span>
                                            model
                                        </p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            To apply M
                                            <span class="s4">4</span>
                                            to some particular forecast problem (i.e., SFP; see Section 3.2.1), the user begins by building a model (see above). To do this, the user needs to modify, in general, only two input text files as described in detail in Section 4.2.2 above:
                                        </p>
                                        <ul id="l46">
                                            <li data-list-text="">
                                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    The build-phase input data file, MMPEInputData_ModelBuildingMode.txt. Examples are provided in the M
                                                    <span class="s4">4</span>
                                                    distribution; one of these, for the Gila River, is shown and described in detail in Section 4.2.2 above. An easy way to do this is to compile the datasets you want to use in a spreadsheet program like Excel for example, using the format shown in the example, and then save it as a tab-delimited text file.
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">The run-control file, MMPE_RunControlFile.txt. For the basic-level WSF user, almost nothing needs to be set in the run control file for the build-phase run. The values and flags as set in the run control file provided in the installation package, and as shown above in Section 4.2.2, are defaults which have been tested for general applicability to the NRCS WSF canonical use-case at a prototype level and do not need to be changed (for basic application) from one SFP to the next. The user need only ensure that, near the top of the file, RunTypeFlag is set to “BUILD”. This can be done in any text editor, for example. Note also that, as mentioned above, the values and flags in the third and final section of the run control file, where the user provides information about how they would like a forecast run to proceed, must be provided during the build phase but their actual values are irrelevant.</p>
                                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">Note that (as mentioned previously) these filenames must not be changed.</p>
                                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                    To run M
                                                    <span class="s4">4</span>
                                                    in RStudio in a Windows environment, simply navigate in Windows Explorer to the subdirectory in which M
                                                    <span class="s4">4</span>
                                                    was installed, double-click on the main program file (MMPE-Main_MkII.R) to open it in RStudio, place the cursor in the upper-left RStudio subwindow which contains the main program script, hit Ctrl-A to highlight the script, and press the Run button at the top right of that window. As noted previously, for the default configuration as applied to a canonical western US statistical WSF use-case as described above, a build-mode run will in general take about 15-35 minutes or so, depending on the computer and other factors. (This run time can be greatly reduced by modifying the genetic algorithm parameters or using manual feature selection instead of the genetic algorithm, but doing so falls into the intermediate-level user instructions provided in Section 4.3.2 below.)
                                                </p>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.3.1.2">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Using the finished model to make a forecast</p>
                                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            To take the M
                                            <span class="s4">4</span>
                                            model suite built in accordance with Section 4.3.1.1 and run it with a new sample to make a forecast as per the usual M
                                            <span class="s4">4</span>
                                            workflow (see background provided earlier in this manual), the user again needs to modify, in general, only two input text files as described in detail in Section 4.2.2 above:
                                        </p>
                                        <ul id="l47">
                                            <li data-list-text="">
                                                <p style="padding-top: 3pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    The forecasting-phase input data file, MMPEInputData_Forecasting.txt. Examples are provided in the M
                                                    <span class="s4">4</span>
                                                    distribution; one of these, for the Gila River, is shown and described in detail in Section
                                                </p>
                                                <p style="padding-left: 43pt;text-indent: 0pt;text-align: left;">4.2.2 above.</p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-top: 1pt;padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">The run-control file, MMPE_RunControlFile.txt. The basic-level WSF user should keep most of the values and flags at the values provided in the run control file provided in the installation package, and as shown above in Section 4.2.2, which are defaults that have been tested for general applicability to the NRCS WSF canonical use-case at a prototype level as discussed above and in the Appendices. However, the user needs to make three sets of modifications, which can be done in any text editor for example:</p>
                                                <ul id="l48">
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 80pt;text-indent: -18pt;line-height: 14pt;text-align: left;">Near the top of the file, RunTypeFlag must be set to “FORECAST”</p>
                                                    </li>
                                                    <li data-list-text="o">
                                                        <p style="padding-left: 80pt;text-indent: -18pt;text-align: left;">In the “RUN PARAMETERS FOR FORECASTING” section near the bottom of the file, the user must specify:</p>
                                                        <ul id="l49">
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: left;">Which PCA modes to retain for each of the models, i.e., the user must provide values for PCSelection_Frwrd_LR, PCSelection_Frwrd_QR, PCSelection_Frwrd_mANN, PCSelection_Frwrd_MCQRNN, PCSelection_Frwrd_RF, and PCSelection_Frwrd_SVM, as described in the detailed summary of MMPE_RunControlFile.txt provided in Sections 4.2.2 above, with further elaboration in Section 4.2.3.3</p>
                                                            </li>
                                                            <li data-list-text="">
                                                                <p style="padding-left: 116pt;text-indent: -18pt;line-height: 108%;text-align: left;">Which predictor variables to retain for each of the models, i.e., the user most provide values for VariableSelection_Frwrd_LR, VariableSelection_Frwrd_QR, VariableSelection_Frwrd_mANN, VariableSelection_Frwrd_MCQRNN, VariableSelection_Frwrd_RF, and VariableSelection_Frwrd_SVM, as described in the detailed summary of MMPE_RunControlFile.txt provided in Sections 4.2.2 above, with further elaboration in Section 4.2.3.3.</p>
                                                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    Let’s look at an example of how to set the values of PCSelection_Frwrd_MCQRNN and VariableSelection_Frwrd_MCQRNN for a forecast run. As explained previously, this information is found in the genetic algorithm’s final output file for the monotone composite quantile regression neural network, GA_RunSummary_PCMCQRNN.txt, written by M
                                                                    <span class="s4">4</span>
                                                                    during the build-phase run. Say that the file looks like this for the Gila River application considered earlier in this manual:
                                                                </p>
                                                                <p class="s18" style="padding-top: 8pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">GA Settings</p>
                                                                <p class="s18" style="padding-left: 54pt;text-indent: 0pt;text-align: left;">Type        = binary chromosome</p>
                                                                <p class="s18" style="padding-left: 54pt;text-indent: 0pt;line-height: 108%;text-align: left;">Population size   = 25 Number of Generations = 10 Elitism       = TRUE</p>
                                                                <p class="s18" style="padding-left: 54pt;text-indent: 0pt;line-height: 10pt;text-align: left;">Mutation Chance   = 0.05</p>
                                                                <p class="s18" style="padding-top: 8pt;padding-left: 54pt;text-indent: -10pt;line-height: 108%;text-align: left;">Search Domain Var 1 = [,]</p>
                                                                <p class="s18" style="padding-left: 54pt;text-indent: 0pt;line-height: 10pt;text-align: left;">Var 0 = [,]</p>
                                                                <p class="s18" style="padding-top: 8pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">GA Results</p>
                                                                <p class="s18" style="padding-left: 54pt;text-indent: 0pt;text-align: left;">Best Solution : 0 1 0 1 0 1 0</p>
                                                                <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    Focus on the string of binary digits in the last line, which encodes the genetic algorithm’s conclusions about how to optimize the features for monotone composite quantile regression neural network. The user provided
                                                                    <i>M</i>
                                                                    = 6 candidate predictors in MMPEInputData_ModelBuildingMode.txt during the build step, so the first 6 binary digits in this string, going from left to right, tell us whether to keep (1) or not
                                                                </p>
                                                                <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">(0) the six predictors that were listed, again from left to right, in MMPEInputData_ModelBuildingMode.txt. All we have to do is copy-and-paste those first six digits into the slot for VariableSelection_Frwrd_MCQRNN in MMPE_RunControlFile.txt:</p>
                                                                <p class="s18" style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">&quot;0 1 0 1 0 1&quot;    # VariableSelection_Frwrd_MCQRNN  - For MCQRNN</p>
                                                                <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    With the default values used in the basic-level use instructions, MaxModes = 2. Recalling from earlier in this manual that the leading PCA mode is always retained, we need only to look at the (
                                                                    <i>M</i>
                                                                    +1)
                                                                    <span class="s4">th</span>
                                                                    (here, seventh) digit in the binary string and see whether it’s a 1 (keep the second PCA mode) or 2 (don’t keep the second PCA mode). In this particular example, it’s 0, so that’s what we enter in the slot for PCSelection_Frwrd_MCQRNN in MMPE_RunControlFile.txt:
                                                                </p>
                                                                <p class="s18" style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">1     # PCSelection_Frwrd_MCQRNN  - For PCMCQRNN</p>
                                                                <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">If that seventh digit in the binary string had happened to be 1 instead, we would’ve entered this in the slot for PCSelection_Frwrd_MCQRNN in MMPE_RunControlFile.txt:</p>
                                                                <p class="s18" style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">1,2    # PCSelection_Frwrd_MCQRNN  - For PCMCQRNN</p>
                                                                <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    The same process needs to be done for the other five modeling systems in M
                                                                    <span class="s4">4</span>
                                                                    , drawing information from their corresponding build-phase genetic algorithm run summary files.
                                                                </p>
                                                                <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                                                    As in the build run, the filenames must not be changed. The process for running M
                                                                    <span class="s4">4</span>
                                                                    in RStudio is described in Section 4.3.1.2 above and is identical for the build and forecast runs. For the default configuration as applied to a canonical western US statistical WSF use-case as described above in this manual, a forecast-mode run will in general take &lt; 1 minute.
                                                                </p>
                                                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">A quick and easy check on whether MMPE_RunControlFile.txt and MMPEInputData_ForecastingMode.txt have been set up correctly for the forecast run is as follows:</p>
                                                            </li>
                                                        </ul>
                                                    </li>
                                                </ul>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-top: 7pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: justify;">Create a test version of MMPEInputData_ForecastingMode.txt that contains the input data from one of the years during the training period, i.e., copying-and-pasting the predictor data from any selected line of MMPEInputData_ModelBuildingMode.txt.</p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-left: 43pt;text-indent: -18pt;text-align: justify;">
                                                    Run M
                                                    <span class="s4">4</span>
                                                    in forecast mode.
                                                </p>
                                            </li>
                                            <li data-list-text="">
                                                <p style="padding-top: 1pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                                    Check that the forecast distributions from each of the constituent modeling systems in M
                                                    <span class="s4">4</span>
                                                    (i.e., as listed in ForwardRunPredictions.csv forecast-mode output file) match the hindcast predictions for that same year for those same models in the M
                                                    <span class="s4">4</span>
                                                    build-phase outputs (i.e., as collectively listed in ensemble_predictions.csv, PCANNpredictions_BC.csv, MCQRNN_predictions.csv, PCQR_predictions.csv, PCR_predictions_BC.csv, PCRF_predictions_BC.csv, and PCSVM_predictions_BC.csv for basic-level users employing the default automatic ensemble composition).
                                                </p>
                                            </li>
                                        </ul>
                                    </li>
                                    <li data-list-text="4.3.1.3">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Other considerations</p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Both the build and forecast phase must take place in the same subdirectory, because forecast runs need to read information from the output files created by M
                                            <span class="s4">4</span>
                                            during the build step.
                                        </p>
                                        <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Note that two identical training runs of M
                                            <span class="s4">4</span>
                                            , using exactly the same dataset and run controls, will in general lead to slightly different results. This may be unfamiliar to users accustomed to linear statistical WSF models, though it is common in automated parameter estimation for process-based hydrologic
                                        </p>
                                        <p style="padding-top: 3pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            models. In M
                                            <span class="s4">4</span>
                                            , this behavior has two main sources. One is the genetic algorithm, which is a stochastic global optimization approach. Without going into detail on the topic (there is plenty of literature on the subject elsewhere) it uses both a gradual march toward better selections of retained input variables and PCA modes based on how the last iteration of the algorithm did, as well as random jumps to help ensure that it finds the true optimum across all possible combinations of candidate input variables and PCA modes without getting stuck in local minima in the hyperdimensional error surface. This can be important for nonlinear, noisy, and high-dimensional problems. Because of the randomized portion of the process, the final result may not be identical from run to run. The other main source of run-to-run variability in build-phase outcomes is that portions of the training process for the various constituent machine learning algorithms also have a stochastic component; for example, initial guesses for neural network weights and biases are typically randomized. Overall, the variation in the final ensemble mean M
                                            <span class="s4">4</span>
                                            WSF from one build run to the next is, in general, slight and of little practical consequence; indeed, it will not be noticed unless multiple identical M
                                            <span class="s4">4</span>
                                            training runs are performed, which would likely only occur if some other issue with the training run was encountered. Multi-model ensembles also tend to damp out the practical impacts of those effects on the final predictions. Steps like increasing the population size and number of generations in the genetic algorithm, or increasing the number of bootstraps in the neural networks, can help mitigate the issue if it is felt to be a problem and will typically lead to incremental prediction improvements more generally, but as discussed previously, this will come at the expense of increased computational times.
                                        </p>
                                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Note that this run-to-run variation does not occur in forecast runs using a given finalized model. That is, taking a M
                                            <span class="s4">4</span>
                                            model suite developed as per section 4.3.1.1 and running it in forecast mode as per section
                                        </p>
                                        <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">4.3.1.2 over and over again using the same forecast-phase input data will yield exactly the same forecast each time (though of course there is no reason to actually do this in practice).</p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Finally, we note that the simple, quick, but nitpicky procedure of manually taking genetic algorithm results from the build-step output files and entering them into the run control file for forecasting runs could be automated with some additional M
                                            <span class="s4">4</span>
                                            code fairly easily. It was left this way, though, because it was known that M
                                            <span class="s4">4</span>
                                            would ultimately be integrated into an operational platform and user interface, in which that functionality could be implemented alongside additional UI-driven features like options for post-hoc user-adjusted ensemble compositions.
                                        </p>
                                    </li>
                                </ol>
                            </li>
                            <li data-list-text="4.3.2">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Intermediate instructions</p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    Intermediate instructions are intended for users who, in addition to the hydrologic expertise described in 4.1 above, also have a good understanding of the principles and approaches of the technical methods employed in M
                                    <span class="s4">4</span>
                                    , though not necessarily in-depth knowledge of the specific algorithms and R packages used to implement those methods.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The intermediate-level application involves user selection, based on professional expertise, of a few non-default options in the run control file, MMPE_RunControlFile.txt, during M
                                    <span class="s4">4</span>
                                    training. Intermediate- level applications should cover, at a prototype level, most if not all of the practical situations arising in contemporary operational WSF as routinely undertaken by service-delivery organizations in the western US and by applied water resource science and engineering researchers working to improve the methods used by those service-delivery organizations.
                                </p>
                                <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">The intermediate-level instructions are identical to the basic-level instructions but accommodate a few additional possibilities.</p>
                                <ol id="l50">
                                    <li data-list-text="4.3.2.1">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Modify the genetic algorithm parameters</p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">As discussed in Section 4.2.2, the larger the population size and the more generations used, the better (roughly speaking, and up to a point) the result of the feature optimization process, but the longer the build run. Experimentation with western US statistical WSF canonical use-cases suggests that a point of diminishing returns is approached relatively quickly. GAPopSize = 25 with GANumGens = 10 appears to be a good compromise and is the recommended default.</p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            However, other choices can be useful in some cases. if the user is in a rush to develop a M
                                            <span class="s4">4</span>
                                            modeling suite for some SFP, GAPopSize = 15 with GANumGens = 7 has generally been found, at a prototype level, to give reasonable results with a shorter run time. Conversely, if the size of the candidate predictor pool is larger, or if build-phase run time is not an issue, it may be a good idea to run a larger genetic algorithm problem size to help ensure that the global best estimate is more closely approximated in the final modeling suite. GAPopSize = 100 with GANumGens = 15, for example, can take several hours to complete on an otherwise identical M
                                            <span class="s4">4</span>
                                            run that would complete in perhaps 15-35 minutes for GAPopSize = 25 with GANumGens = 10, or even less for GAPopSize = 15 with GANumGens = 7, but will more reliably obtain a truly optimal feature selection for a canonical WSF use-case scenario, though in testing so far the improvement will not be dramatic relative to the default genetic algorithm settings for the canonical WSF use cases we focus on here.
                                        </p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">Note that GAPopSize and GANumGens can be varied independently of each other by the user, and similarly, can be varied independently of other run control parameters.</p>
                                    </li>
                                    <li data-list-text="4.3.2.2">
                                        <p class="s14" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Retain fewer or more PCA modes</p>
                                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">The default case is MaxModes = 2, that is, the genetic algorithm has to decide whether or not to include the second PCA mode. This is again a compromise solution.</p>
                                        <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">Physical interpretation is one issue to consider in WSF applications. For canonical use-cases where the main predictive inputs are variables like current SWE and accumulated precipitation, the leading PCA mode effectively constitutes a watershed-scale index of wintertime hydroclimatic conditions, and in practice often captures 80% or more of the total input dataset variance. In a substantial minority of existing NRCS PCR-based WSF models, though, antecedent streamflow is included in the predictor candidate pool to index river flow inputs from natural water storages in a watershed, like soil moisture, groundwater, or wetlands; these geophysical processes are obviously influenced by, but physically distinct from, SWE and precipitation and may appear as a distinct second PCA mode that will in general be retained by a feature optimization routine and has clear physical meaning. Even if such distinct input variables are not included in the candidate pool, more than one PCA mode may be retained in some cases, presumably reflecting second-order spatiotemporal patterns in snowpack or precipitation accumulation across the watershed.</p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;text-align: left;">That said, there may be cases where the hydrologist wishes to consider fewer or more PCA modes.</p>
                                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">The case for fewer modes (that is, MaxModes = 1) is simple: in the canonical use-cases described above the leading mode is usually the strongly dominant and most physically explainable PCA mode; this</p>
                                        <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">choice reduces the amount of work that the genetic algorithm has to do; and it also makes the forward run easier to set up in the run control file, as PCSelection_Frwrd_LR, PCSelection_Frwrd_QR, PCSelection_Frwrd_mANN, PCSelection_Frwrd_MCQRNN, PCSelection_Frwrd_RF, and PCSelection_Frwrd_SVM are all simply 1 in this case. The trade-off is a potential loss of some predictive skill, and possibly loss of some physical insight if patterns in the predictors chosen for the candidate pool can be legitimately decomposed by PCA into more than one significant geophysical process.</p>
                                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">Conversely, testing so far suggests it’s not unusual in canonical use-case scenarios to be able to eke out a little more predictive skill by increasing MaxModes from its default value to 3 or 4. The trade-offs for this choice in the canonical use-cases are that it may be difficult to reliably attribute physical meaning to the third or fourth PCA mode; these higher modes typically explain very little of the total input dataset variance; and the scale of the optimization problem may be substantially increased, with potential implications for run times or global optimality of the solution. It also makes setting PCSelection_Frwrd_LR, PCSelection_Frwrd_QR, PCSelection_Frwrd_mANN, PCSelection_Frwrd_MCQRNN, PCSelection_Frwrd_RF, and PCSelection_Frwrd_SVM more complicated (see above detailed instructions, such as in Sections 4.2.2 and 4.2.3.3).</p>
                                        <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">Note that MaxModes can be varied independently of other run control parameters.</p>
                                    </li>
                                    <li data-list-text="4.3.2.3">
                                        <p class="s14" style="padding-top: 9pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Turn off the genetic algorithm</p>
                                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">It may be useful to turn off automated feature optimization in some cases. The user may wish all variables in MMPEInputData_ModelBuildingMode.txt to be used, perhaps for testing purposes or because the user is confident that all the predictor variables provided in the input data file should be included in the model, perhaps for geophysical or operational redundancy reasons. Another possibility is that very few input variables are available; in this case, the genetic algorithm may not have much to do. Speediness is another factor; a build-phase run that could take roughly 15-35 minutes with feature optimization using default genetic algorithm parameters may run in &lt; 1 minute if feature optimization is switched off.</p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">To do this, in the run control file, during the build run set GeneticAlgorithmFlag to “N” and ManualPCSelection to whichever PCA modes the user wishes to retain (e.g., 1 or 1,2 or 1,3,4 etc.) (see detailed instructions in Section 4.2.2). Generally speaking, all other flags and values in MMPE_RunControlFile.txt can be left at their default settings.</p>
                                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">For the forecast run, in the run control file, just copy-and-paste the values set by the user for ManualPCSelection during the build run into the slots for which PCA modes to retain in the forward run. For instance, if the user specified retention of the first and second PCA modes in training:</p>
                                        <p class="s18" style="padding-top: 8pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">1,2     # ManualPCSelection</p>
                                        <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">then use the following in MMPE_RunControlFile.txt for the forecast run:</p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <table style="border-collapse:collapse;margin-left:41.5pt" cellspacing="0">
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCSelection_Frwrd_LR</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCSelection_Frwrd_QR</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCQR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCSelection_Frwrd_mANN</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCANN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCSelection_Frwrd_MCQRNN</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCMCQRNN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCSelection_Frwrd_RF</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCRF</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:54pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 9pt;text-align: left;">1,2</p>
                                                </td>
                                                <td style="width:43pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 9pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:140pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 9pt;text-align: left;">PCSelection_Frwrd_SVM</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 9pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:49pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 9pt;text-align: left;">PCSVM</p>
                                                </td>
                                            </tr>
                                        </table>
                                        <p style="padding-top: 2pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                            Analogously, for the forecast run, set the input variable selections to a length-
                                            <i>M</i>
                                            string of 1s. For example, if the MMPEInputData_ModelBuildingMode.txt file supplied by the user for the build run (and therefore the MMPEInputData_ForecastingMode.txt file as well) contains six variables as in the Gila case shown above in this manual, then in MMPE_RunControlFile.txt set these selections to a string of six 1s separated by a single space with quotation marks at the start and end of the string:
                                        </p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <table style="border-collapse:collapse;margin-left:41.5pt" cellspacing="0">
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 10pt;text-align: left;"># VariableSelection_Frwrd_LR</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 10pt;text-align: left;"># VariableSelection_Frwrd_QR</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCQR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 10pt;text-align: left;"># VariableSelection_Frwrd_mANN</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCANN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 10pt;text-align: left;"># VariableSelection_Frwrd_MCQRNN</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">MCQRNN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 10pt;text-align: left;"># VariableSelection_Frwrd_RF</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 10pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCRF</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:100pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 9pt;text-align: left;">&quot;1 1 1 1 1 1&quot;</p>
                                                </td>
                                                <td style="width:208pt">
                                                    <p class="s19" style="padding-left: 26pt;text-indent: 0pt;line-height: 9pt;text-align: left;"># VariableSelection_Frwrd_SVM</p>
                                                </td>
                                                <td style="width:37pt">
                                                    <p class="s19" style="padding-right: 2pt;text-indent: 0pt;line-height: 9pt;text-align: right;">- For</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-left: 2pt;text-indent: 0pt;line-height: 9pt;text-align: left;">PCSVM</p>
                                                </td>
                                            </tr>
                                        </table>
                                    </li>
                                    <li data-list-text="4.3.2.4">
                                        <p class="s14" style="padding-top: 8pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Single predictor variable</p>
                                        <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">This is merely a special case of Section 4.3.2.3 in which MMPEInputData_ModelBuildingMode.txt contains only one predictor variable. That also means that only one PCA mode is available. (Of course, no PCA is actually required if only a single variable exists, but because this is an edge case, rather than coding a separate workflow the single predictor is simply routed through the normal PCA workflow.) In this case, during the build phase use the following in MMPEInputData_ModelBuildingMode.txt:</p>
                                        <p class="s18" style="padding-top: 7pt;padding-left: 44pt;text-indent: 0pt;text-align: left;">&quot;N&quot;     # GeneticAlgorithmFlag</p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">and:</p>
                                        <p class="s18" style="padding-top: 9pt;padding-left: 49pt;text-indent: 0pt;text-align: left;">1     # ManualPCSelection</p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">Then in the forecasting step, use the following:</p>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                        <table style="border-collapse:collapse;margin-left:5.5pt" cellspacing="0">
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">PCSelection_Frwrd_LR   - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">PCSelection_Frwrd_QR   - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCQR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">PCSelection_Frwrd_mANN  - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCANN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">PCSelection_Frwrd_MCQRNN - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCMCQRNN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: right;">PCSelection_Frwrd_RF   - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;line-height: 10pt;text-align: left;">PCRF</p>
                                                </td>
                                            </tr>
                                            <tr style="height:17pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="text-indent: 0pt;text-align: center;">1</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="text-indent: 0pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;text-align: right;">PCSelection_Frwrd_SVM  - For</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-left: 5pt;text-indent: 0pt;text-align: left;">PCSVM</p>
                                                </td>
                                            </tr>
                                            <tr style="height:22pt">
                                                <td style="width:82pt">
                                                    <p class="s20" style="padding-top: 3pt;padding-left: 2pt;text-indent: 0pt;text-align: left;">and:</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p style="text-indent: 0pt;text-align: left;">
                                                        <br>
                                                    </p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p style="text-indent: 0pt;text-align: left;">
                                                        <br>
                                                    </p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p style="text-indent: 0pt;text-align: left;">
                                                        <br>
                                                    </p>
                                                </td>
                                            </tr>
                                            <tr style="height:16pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-top: 4pt;padding-right: 26pt;text-indent: 0pt;line-height: 10pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-top: 4pt;padding-right: 5pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="padding-top: 4pt;text-indent: 0pt;line-height: 10pt;text-align: left;">VariableSelection_Frwrd_LR</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="padding-top: 4pt;text-indent: 0pt;line-height: 10pt;text-align: left;">- For PCR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-right: 26pt;text-indent: 0pt;line-height: 10pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 5pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">VariableSelection_Frwrd_QR</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">- For PCQR</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-right: 26pt;text-indent: 0pt;line-height: 10pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 5pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">VariableSelection_Frwrd_mANN</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">- For PCANN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-right: 26pt;text-indent: 0pt;line-height: 10pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 5pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">VariableSelection_Frwrd_MCQRNN</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">- For MCQRNN</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-right: 26pt;text-indent: 0pt;line-height: 10pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 5pt;text-indent: 0pt;line-height: 10pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">VariableSelection_Frwrd_RF</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 10pt;text-align: left;">- For PCRF</p>
                                                </td>
                                            </tr>
                                            <tr style="height:11pt">
                                                <td style="width:82pt">
                                                    <p class="s19" style="padding-right: 26pt;text-indent: 0pt;line-height: 9pt;text-align: right;">&quot;1&quot;</p>
                                                </td>
                                                <td style="width:38pt">
                                                    <p class="s19" style="padding-right: 5pt;text-indent: 0pt;line-height: 9pt;text-align: right;">#</p>
                                                </td>
                                                <td style="width:178pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 9pt;text-align: left;">VariableSelection_Frwrd_SVM</p>
                                                </td>
                                                <td style="width:67pt">
                                                    <p class="s19" style="text-indent: 0pt;line-height: 9pt;text-align: left;">- For PCSVM</p>
                                                </td>
                                            </tr>
                                        </table>
                                        <p style="text-indent: 0pt;text-align: left;">
                                            <br>
                                        </p>
                                    </li>
                                </ol>
                            </li>
                            <li data-list-text="4.3.3">
                                <p class="s5" style="padding-left: 44pt;text-indent: -36pt;text-align: left;">Expert instructions</p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    In addition to hydrologic expertise and solid understanding of the principles of the techniques deployed in M
                                    <span class="s4">4</span>
                                    as described in 4.2 above, expert users have an in-depth understanding of the M
                                    <span class="s4">4</span>
                                    code itself, the various R packages it incorporates, and the underlying theory and applications around the algorithms used by M
                                    <span class="s4">4</span>
                                    and the R packages it employs.
                                </p>
                                <p style="padding-top: 3pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    As noted above, the basic-level application relies heavily on AutoML, which in M
                                    <span class="s4">4</span>
                                    essentially amounts to a combination of custom automation algorithms and carefully selected and tested default hyperparameter choices. This remains largely true of the intermediate-level application as well, which for the most part merely changes some hyperhyperparameters or other predefined options in the run control file.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">In contrast, expert users access various additional capabilities, such as:</p>
                                <ul id="l51">
                                    <li data-list-text="">
                                        <p style="padding-top: 9pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">Modifying the values of additional flags and variables in the run control file beyond the specific cases described in section 4.3.2.</p>
                                    </li>
                                    <li data-list-text="">
                                        <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            Modifying additional hyperparameters and flags within the M
                                            <span class="s4">4</span>
                                            calls to the R packages through small edits to the M
                                            <span class="s4">4</span>
                                            code. One example would be to go into the custom R function file implementing the monotone artificial neural network (PCANN-Module_MkII.R; see Section 4.2.1) and manually increase the number of hidden-layer neurons in the artificial neural network beyond the possibilities considered in the M
                                            <span class="s4">4</span>
                                            AutoML routine designed for this task; this could conceivably be required if larger or more diverse predictor datasets are used relative to the canonical or near-canonical use-cases described above in 4.3.1 and 4.3.2; indeed, even a light version of deep learning, i.e., a neural network with more than one hidden layer and many neurons in those hidden layers, could be implemented in this way.
                                        </p>
                                    </li>
                                    <li data-list-text="">
                                        <p style="padding-left: 44pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                                            More extensively modify the open-source M
                                            <span class="s4">4</span>
                                            code itself; as just one of many possible examples, by adding a seventh forecasting model of their choice to the metasystem.
                                        </p>
                                    </li>
                                </ul>
                                <p style="text-indent: 0pt;text-align: left;">
                                    <br>
                                </p>
                            </li>
                        </ol>
                    </li>
                </ol>
            </li>
            <li data-list-text="5">
                <h3 style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">
                    Files provided in the M
                    <span class="s3">4</span>
                    distribution
                </h3>
                <p style="text-indent: 0pt;text-align: left;">
                    <br>
                </p>
                <ol id="l52">
                    <li data-list-text="5.1">
                        <h4 style="padding-top: 15pt;padding-left: 43pt;text-indent: -36pt;text-align: left;">
                            M
                            <span class="s7">4</span>
                            installation files
                        </h4>
                        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            The M
                            <span class="s4">4</span>
                            code files were described above in Section 4.2.1, which can be installed as per the instructions given in that section.
                        </p>
                        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            One example of each of the three required M
                            <span class="s4">4</span>
                            input files described above in Section 4.2.2 (MMPE_RunControlFile.txt, MMPEInputData_ModelBuildingMode.txt, and MMPEInputData_ForecastingMode.txt) is also provided along with the code because these need to be in quite specific formats; these examples therefore provide a template for the user to employ when setting up the model for a given SFP. The run control file given here contains default flags and values as described in detail in Section 4.2.2, though any forecast-mode run of M
                            <span class="s4">4</span>
                            will require the user to set a few values manually, also as described above. As a FYI, the two input data files happen to correspond to an April forecast of April-July flow volume of the Yellowstone River at Corwin Springs, using 30 years of data for training and 18 predictor variates corresponding to NRCS SNOTEL and snow course sites.
                        </p>
                        <p style="text-indent: 0pt;text-align: left;">
                            <br>
                        </p>
                    </li>
                    <li data-list-text="5.2">
                        <h4 style="padding-left: 43pt;text-indent: -36pt;text-align: left;">Completed examples</h4>
                        <p style="padding-top: 3pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                            The distribution also includes four completed examples of how to run M
                            <span class="s4">4</span>
                            , each in a different folder. Each folder contains the source code, input files, and build-step and forecast-step output files for a different M
                            <span class="s4">4</span>
                            application. The four examples are described below.
                        </p>
                        <ol id="l53">
                            <li data-list-text="5.2.1">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;line-height: 108%;text-align: left;">Example 1: Gila River at Gila, March 1 forecast of March-through-May volume, genetic algorithm feature selection with up to 2 PCA modes</p>
                                <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">This is an example of the default, basic-level use instructions in Section 4.3.1 above.</p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">The model-building step followed the workflow described in Section 4.3.1.1: (a) created the input data file, MMPEInputData_ModelBuildingMode.txt, containing historical flow volume data and a pool of candidate predictor variables, which in this example was the Gila River data file shown above in Section</p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    4.2.2; (b) made sure that RunTypeFlag was set to “BUILD” in MMPE_RunControlFile.txt, left everything else in that file at its default values as shown in Section 4.2.2, and saved and closed it; (c) ran M
                                    <span class="s4">4</span>
                                    in RStudio. When complete, M
                                    <span class="s4">4</span>
                                    wrote the large number of output files listed in Sections 4.2.3.2 and 4.2.3.3.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    The forecasting step followed the workflow described in Section 4.3.1.2: (a) put together a new sample of the predictor set in MMPEInputData_ForecastingMode.txt, which in this example was the Gila River data file shown above in Section 4.2.2; (b) went back into MMPE_RunControlFile.txt, changed RunTypeFlag to “FORECAST”, set the values of PCSelection_Frwrd_LR, PCSelection_Frwrd_QR, PCSelection_Frwrd_mANN, PCSelection_Frwrd_MCQRNN, PCSelection_Frwrd_RF, PCSelection_Frwrd_SVM, VariableSelection_Frwrd_LR, VariableSelection_Frwrd_QR, VariableSelection_Frwrd_mANN, VariableSelection_Frwrd_MCQRNN, VariableSelection_Frwrd_RF, and VariableSelection_Frwrd_SVM, following the instructions provided in Section 4.3.1.2, and saved and closed it again; (c) ran M
                                    <span class="s4">4</span>
                                    in RStudio. This was all done in the same folder as the build step, because the forecast step needs to have the build-step output files available to it, as noted previously. When complete, M
                                    <span class="s4">4</span>
                                    wrote the two output files listed in Section 4.2.3.5.
                                </p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: justify;">
                                    Recall from Section 4.3.1.3 that each time M
                                    <span class="s4">4</span>
                                    is run in training mode, it gets slightly different models at the end due to stochasticity in the genetic algorithm (e.g., random mutations) and some of the machine learning training processes (e.g., random initial weights and biases in an ANN). Refer back to Section
                                </p>
                                <p style="padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
                                    4.3.1.3 above for details. In practice, this means that even if you run M
                                    <span class="s4">4</span>
                                    in build mode on exactly the same training dataset with exactly the same run control file as in Example 1, the final result (potentially including which input variables and PCA modes to retain) will in general show some slight differences from the Example 1 output files posted to the GitHub repo. Consequently, if you want to take those new models you just created and use them in a forecast run, you’ll have the follow the steps laid out above – that is, in addition to changing “BUILD” to “FORECAST” in the run control file, you’ll also have to set the PCSelection_Frwrd_(modelname) and VariableSelection_Frwrd_(modelname) values in accordance with the outcomes from your new build-phase M
                                    <span class="s4">4</span>
                                    run. Conversely, if you just want to run M
                                    <span class="s4">4</span>
                                    in forecast mode using the Example 1 training outputs posted to the GitHub repo, without retraining the model, no changes to the run control file would be needed.
                                </p>
                            </li>
                            <li data-list-text="5.2.2">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;line-height: 108%;text-align: left;">Example 2: Gila River at Gila, March 1 forecast of March-through-May volume, genetic algorithm feature selection with up to 4 PCA modes</p>
                                <p style="padding-top: 2pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">This is an example of the intermediate-level use instructions in Section 4.3.2.2. That is, this is the same problem as in Example 1 and described in Section 5.2.1 above, following exactly the same instructions, except that the genetic algorithm was instructed to retain up to a maximum of four PCA modes by setting MaxModes to 4 in MMPE_RunControlFile.txt during the build phase, instead of the default value of 2. Nothing else was different.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">In practice, one would rarely want to set MaxModes to anything over 2 for the western US canonical statistical WSF use-case, but other situations could conceivably arise where more PCA modes are needed. (The situation where MaxModes is decreased from the default of 2 down to 1 is a much more realistic scenario; see Section 4.3.2.2 above.)</p>
                            </li>
                            <li data-list-text="5.2.3">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;line-height: 108%;text-align: left;">Example 3: Owyhee River at Rome, April 1 forecast of April-through-July volume, genetic algorithm feature selection with up to 2 PCA modes</p>
                                <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">This is another example of the default, basic-level use instructions in Section 4.3.1 above. It is the identical to the example of Section 5.2.1, but for a different river and so obviously uses different observed flow and predictor data in MMPEInputData_ModelBuildingMode.txt and MMPEInputData_ForecastingMode.txt.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">In this case, the candidate predictor pool happens to be much larger than for the Gila. This example also demonstrates what the results look like when the nonnegativity check algorithm removes some models as it finalizes the multi-model ensemble (in this case, linear regression and linear quantile regression were removed).</p>
                            </li>
                            <li data-list-text="5.2.4">
                                <p class="s5" style="padding-top: 7pt;padding-left: 44pt;text-indent: -36pt;line-height: 108%;text-align: left;">Example 4: Deschutes River below Snow Creek, February 1 forecast of April-through- July volume, genetic algorithm switched off</p>
                                <p style="padding-top: 8pt;padding-left: 8pt;text-indent: 0pt;text-align: left;">This is an example of the intermediate-level use instructions in Section 4.3.2.3 above.</p>
                                <p style="padding-top: 9pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">That is, this example demonstrates the use of manual feature selection, in which the genetic algorithm is turned off by setting GeneticAlgorithmFlag to “N” and (in this particular example) ManualPCSelection to 1,2 such that the leading and second PCA modes are used for all six constituent modeling systems. All predictors in the input data files, MMPEInputData_ModelBuildingMode.txt and MMPEInputData_ForecastingMode.txt, are used in all six constituent modeling systems.</p>
                                <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">This example also demonstrates how predictors other than SWE and precipitation are included in the input data files. In this particular case, antecedent streamflow at a location on the Deschutes River is included as the right-most predictor variable in the input data matrix (note that it could, of course, be placed anywhere among the predictor variables, not necessarily on the far right-hand side). It also illustrates an early-season forecast, which in general has substantially lower predictive skill than forecasts later in the season. The inclusion of antecedent streamflow and the fact that it’s an early- season forecast don’t have anything to do with any of the settings in MMPE_RunControlFile.txt, however, or vice versa; for instance, exactly the same input data file, containing antecedent streamflow, could be used when applying the default, basic-level use instructions as described in Section 4.3.1 above.</p>
                            </li>
                        </ol>
                    </li>
                </ol>
            </li>
            <li data-list-text="6">
                <h3 style="padding-top: 1pt;padding-left: 44pt;text-indent: -36pt;text-align: left;">Known issues and bug fixes</h3>
            </li>
        </ol>
        <p style="padding-top: 9pt;padding-left: 8pt;text-indent: 0pt;line-height: 108%;text-align: left;">
            Tracking of issues and bug fixes in this document and on the GitHub site begins with the first GitHub distribution of M
            <span class="s4">4</span>
            In March 2023, and which consisted of the following files:
        </p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">MMPE-Main_MkII.R PCA-Module_v5.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCA-Graphics-Module_v3.R</p>
        <p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">GA-PredictorSelection-Module_v15.R</p>
		<p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCR-Module_v11.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCSVM-Module_v6.R</p>
		<p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCRF-Module_v3.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCANN-Module_v17.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCQR-Module_v7.R PCMCQRNN_v5.R</p>
		<p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCMCQRNN_v5.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">InitializeEnsemble-Module_v1.R</p>
		<p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">AppendEnsemble-Module_v1.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">FinalizeEnsemble-Module_v1.R</p>
        <p style="padding-top: 7pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">Diagnostics-Module_v6.R</p>

        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;text-align: left;">
            Known issues with this version of M
            <span class="s4">4</span>
            are:
        </p>
        <ul id="l54">
            <li data-list-text="">
                <p style="padding-top: 9pt;padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                    There is almost no error-trapping for user mistakes. It is up to the user to understand M
                    <span class="s4">4</span>
                    , correctly set values and flags in the run control file, and correctly provide data in the input files, using the required formatting. Failure to do so may result in M
                    <span class="s4">4</span>
                    outputs different from what the user thought they were asking for but actually didn’t, or outright crashes.
                </p>
            </li>
            <li data-list-text="">
                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                    RStudio can be a little persnickety, occasionally leading to M
                    <span class="s4">4</span>
                    crashes, sometimes in seemingly random ways, that aren’t the fault of either M
                    <span class="s4">4</span>
                    or the user. See Section 4.1 for tips and alternatives.
                </p>
            </li>
            <li data-list-text="">
                <p style="padding-left: 43pt;text-indent: -18pt;line-height: 108%;text-align: left;">
                    As noted in the detailed description given for the run control flag, AutoEnsembleFlag, in Section 4.2.2, the Box-Cox transform space-based prediction intervals used in the default versions of four of the six modeling systems in M
                    <span class="s4">4</span>
                    are generally superior but can occasionally act up. When this happens, the prediction intervals are noticeably too wide. Plans call for this issue to be addressed in a future version of M
                    <span class="s4">4</span>
                    . For now, the user has the option of using non-default versions of one or more of those four modeling systems that employ more conventional arithmetic-space prediction intervals, or if preferred, omitting the misbehaving modeling system altogether from the multi-model ensemble forecast distribution. Refer back to Sections 4.2.2 and 4.2.3 for further details.
                </p>
            </li>
        </ul>
        <p style="padding-left: 44pt;text-indent: -36pt;line-height: 23pt;text-align: left;">
            The current GitHub repo (as of May 2023) for M
            <span class="s4">4</span>
            consists of the following files: MMPE-Main_MkII.R
        </p>
        <p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCA-Module_MkII.R</p>
        <p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCA-Graphics-Module_MkII.R</p>
        <p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">GA-PredictorSelection-Module_MkII.R</p>
        <p style="padding-top: 1pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCR-Module_MkII.R</p>
		<p style="padding-top: 2pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCSVM-Module_MkII.R</p>
        <p style="padding-top: 2pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCRF-Module_MkII.R</p>
		<p style="padding-top: 2pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCANN-Module_MkII.R</p>
		<p style="padding-top: 2pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCQR-Module_MkII.R</p>
		<p style="padding-top: 2pt;padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">PCMCQRNN_MkII.R</p>
		<p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">InitializeEnsemble-Module_MkII.R</p>
        <p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">AppendEnsemble-Module_MkII.R</p>
		<p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">FinalizeEnsemble-Module_MkII.R</p>
		<p style="padding-left: 43pt;text-indent: 0pt;line-height: 108%;text-align: left;">Diagnostics-Module_MkII.R</p>

		<p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
            Two functionality changes were made relative to the previous version: (1) during training runs, in rare severe cases of a misbehaving Box-Cox transform space-based prediction interval leading to Inf values that cause subsequent graphing steps to crash, M
            <span class="s4">4</span>
            automatically overwrites the Box-Cox transform space-based prediction intervals with conventional stationary Gaussian prediction intervals, and lets the user know; and (2) the forecast run produces forecasts and prediction intervals for all 10 models in M
            <span class="s4">4</span>
            . These changes occur in the main program and cause M
            <span class="s4">4</span>
            to generate an additional training-run output file, BoxCoxPredictionIntervalsSwitches.csv, as well as changes to the forecast-run output files, of which there are now two, ForwardRunMultiModelEnsembleMeanPrediction.csv and ForwardRunAllModelsPredictions.csv. See Sections 4.2.2 and 4.2.3 for further details on technical background, and Sections 4.2.3.2 and 4.2.3.5 for further details on the output file changes.
        </p>
        <p style="padding-top: 7pt;padding-left: 7pt;text-indent: 0pt;line-height: 108%;text-align: left;">
            Additionally, the files were renamed for better consistency with GitHub standard practices for version control, and in some sense semantic versioning more generally, mitigating version lock issues for example. However, the custom function files (i.e, all M
            <span class="s4">4</span>
            .R files other than the main program) are otherwise substantially unchanged.
        </p>
        <p style="padding-top: 8pt;padding-left: 7pt;text-indent: 0pt;text-align: left;">Other issues or caveats are as noted above for the original distribution.</p>
    </body>
</html>
